<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Flexible Genotyping for Polyploids • updog</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Flexible Genotyping for Polyploids">
<meta property="og:description" content='Implements empirical Bayes approaches to genotype
       polyploids from next generation sequencing data while
       accounting for allelic bias, overdispersion, and sequencing
       error. The main function is flexdog(), which allows the specification
       of many different genotype distributions. An experimental
       function that takes into account varying levels of relatedness
       is implemented in mupdog(). Also provided are functions to
       simulate genotypes, rgeno(), and read-counts, rflexdog(), as well as
       functions to calculate oracle genotyping error rates, oracle_mis(), and
       correlation with the true genotypes, oracle_cor(). These latter two
       functions are useful for read depth calculations. Run
       browseVignettes(package = "updog") in R for example usage. See also
       Gerard et al. (2018) &lt;doi:10.1534/genetics.118.301468&gt; for details on the
       implemented methods.'>
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="updog" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#updog" class="anchor"></a>updog</h1></div>

<p>Updog provides a suite of methods for genotyping polyploids from next-generation sequencing (NGS) data. It does this while accounting for many common features of NGS data: allele bias, overdispersion, sequencing error, and (possibly) outlying observations. It is named updog for “Using Parental Data for Offspring Genotyping” because we originally developed the method for full-sib populations, but it works now for more general populations. The method is described in detail <a href="https://doi.org/10.1534/genetics.118.301468">here</a> and <a href="https://doi.org/10.1101/281550">here</a>.</p>
<p>The main function is <code><a href="reference/flexdog.html">flexdog()</a></code>, which provides many options for the distribution of the genotypes in your sample. Novel genotype distributions include the class of proportional normal distributions (<code>model = "norm"</code>) and the class of discrete unimodal distributions (<code>model = "ash"</code>). The default is <code>model = "norm"</code> because it is the most robust to varying genotype distributions, but feel free to use more specialized priors if you have more information on the data.</p>
<p>Also provided are:</p>
<ul>
<li>An experimental function <code><a href="reference/mupdog.html">mupdog()</a></code>, which allows for correlation between the individuals’ genotypes while jointly estimating the genotypes of the individuals at all provided SNPs. The implementation uses a variational approximation. This is designed for samples where the individuals share a complex relatedness structure (e.g. siblings, cousins, uncles, half-siblings, etc). Right now there are no guarantees about this function’s performance.</li>
<li>Functions to simulate genotypes (<code><a href="reference/rgeno.html">rgeno()</a></code>) and read-counts (<code><a href="reference/rflexdog.html">rflexdog()</a></code>). These support all of the models available in <code><a href="reference/flexdog.html">flexdog()</a></code>.</li>
<li>Functions to evaluate oracle genotyping performance: <code><a href="reference/oracle_joint.html">oracle_joint()</a></code>, <code><a href="reference/oracle_mis.html">oracle_mis()</a></code>, <code><a href="reference/oracle_mis_vec.html">oracle_mis_vec()</a></code>, and <code><a href="reference/oracle_cor.html">oracle_cor()</a></code>. We mean “oracle” in the sense that we assume that the entire data generation process is known (i.e. the genotype distribution, sequencing error rate, allele bias, and overdispersion are all known). These are good approximations when there are a lot of individuals (but not necessarily large read-depth).</li>
</ul>
<p>The original <code>updog</code> package is now named <code>updogAlpha</code> and may be found <a href="https://github.com/dcgerard/updogAlpha">here</a>.</p>
<p>See also <a href="https://github.com/pblischak/polyploid-genotyping">ebg</a>, <a href="https://cran.r-project.org/package=fitPoly">fitPoly</a>, and <a href="http://www.g3journal.org/content/suppl/2017/01/19/g3.117.039008.DC1">TET</a>, and <a href="https://cran.r-project.org/package=polyRAD">polyRAD</a>. Our best “competitor” is probably <a href="https://cran.r-project.org/package=fitPoly">fitPoly</a>, though <a href="https://cran.r-project.org/package=polyRAD">polyRAD</a> has some nice ideas for utilizing population structure and linkage disequilibrium.</p>
<p>See <a href="./inst/NEWS.html">NEWS</a> for the latest updates on the package.</p>
<div id="vignettes" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h2>
<p>I’ve included many vignettes in <code>updog</code>, which you can access online <a href="https://dcgerard.github.io/updog/">here</a>.</p>
</div>
<div id="bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-reports" class="anchor"></a>Bug Reports</h2>
<p>If you find a bug or want an enhancement, please submit an issue <a href="http://github.com/dcgerard/updog/issues">here</a>.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install updog from <a href="https://cran.r-project.org/package=updog">CRAN</a> in the usual way:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"updog"</span>)</a></code></pre></div>
<p>You can install the current (unstable) version of updog from GitHub with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># install.packages("devtools")</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"dcgerard/updog"</span>)</a></code></pre></div>
<div id="cvxr" class="section level3">
<h3 class="hasAnchor">
<a href="#cvxr" class="anchor"></a>CVXR</h3>
<p>If you want to use the <code>use_cvxr = TRUE</code> option in <code>flexdog</code> (not generally recommended), you will need to install the <a href="https://cran.r-project.org/package=CVXR">CVXR</a> package. Before I could install CVXR in Ubuntu, I had to run in the terminal</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="fu">sudo</span> apt-get install libmpfr-dev</a></code></pre></div>
<p>and then run in R</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"Rmpfr"</span>)</a></code></pre></div>
</div>
</div>
<div id="how-to-cite" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-cite" class="anchor"></a>How to Cite</h2>
<p>Please cite</p>
<blockquote>
<p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. doi: <a href="https://doi.org/10.1534/genetics.118.301468">10.1534/genetics.118.301468</a>.</p>
</blockquote>
<p>Or, using BibTex:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode tex"><code class="sourceCode latex"><a class="sourceLine" id="cb5-1" data-line-number="1">@article {gerard2018genotyping,</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">    author = {Gerard, David and Ferr{<span class="fu">\~</span>a}o, Luis Felipe Ventorim and Garcia, Antonio Augusto Franco and Stephens, Matthew},</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">    title = {Genotyping Polyploids from Messy Sequencing Data},</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">    volume = {210},</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">    number = {3},</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">    pages = {789--807},</a>
<a class="sourceLine" id="cb5-7" data-line-number="7">    year = {2018},</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">    doi = {10.1534/genetics.118.301468},</a>
<a class="sourceLine" id="cb5-9" data-line-number="9">    publisher = {Genetics},</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">    issn = {0016-6731},</a>
<a class="sourceLine" id="cb5-11" data-line-number="11">    URL = {https://doi.org/10.1534/genetics.118.301468},</a>
<a class="sourceLine" id="cb5-12" data-line-number="12">    journal = {Genetics}</a>
<a class="sourceLine" id="cb5-13" data-line-number="13">}</a></code></pre></div>
</div>
<div id="code-of-conduct" class="section level2">
<h2 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of Conduct</h2>
<p>Please note that this project is released with a <a href="https://github.com/dcgerard/updog/blob/master/CONDUCT.md">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=updog">https://​cloud.r-project.org/​package=updog</a>
</li>
<li>Report a bug at <br><a href="http://github.com/dcgerard/updog/issues">http://​github.com/​dcgerard/​updog/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing updog</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>David Gerard <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9450-5023" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/dcgerard/updog"><img src="https://travis-ci.org/dcgerard/updog.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/dcgerard/updog-06s8t"><img src="https://ci.appveyor.com/api/projects/status/c80fdy61ead6s3vr?svg=true" alt="Appveyor Build status"></a></li>
<li><a href="https://codecov.io/gh/dcgerard/updog"><img src="https://codecov.io/gh/dcgerard/updog/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://www.gnu.org/licenses/gpl-3.0"><img src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" alt="License: GPL v3"></a></li>
<li><a href="https://cran.r-project.org/package=updog"><img src="http://www.r-pkg.org/badges/version/updog" alt="CRAN_Status_Badge"></a></li>
<li><a href="https://cran.r-project.org/package=updog"><img src="http://cranlogs.r-pkg.org/badges/grand-total/updog"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
