<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="David Gerard" />

<meta name="date" content="2018-06-11" />

<title>Parallelized Genotyping using Updog</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Parallelized Genotyping using Updog</h1>
<h4 class="author"><em>David Gerard</em></h4>
<h4 class="date"><em>2018-06-11</em></h4>



<div id="abstract" class="section level1">
<h1>Abstract</h1>
<p>We provide two examples of parallelizing <code>flexdog</code> to genotype many genes.</p>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Before we begin, we’ll need to load the <code>parallel</code>, <code>foreach</code>, and <code>doParallel</code> packages. There are other ways to perform parallel computation in R, but we find these to be pretty handy and easy to use.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(parallel)
<span class="kw">library</span>(foreach)
<span class="kw">library</span>(doParallel)</code></pre></div>
<p>Of course, we’ll need to load the <code>updog</code> package to actually do the genotyping.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(updog)</code></pre></div>
<p>We’ll work with the <code>uitedewilligen</code> dataset, which is a subset of the data from <span class="citation">Uitdewilligen (2013)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;uitdewilligen&quot;</span>)
ngenes &lt;-<span class="st"> </span><span class="kw">ncol</span>(uitdewilligen<span class="op">$</span>refmat)</code></pre></div>
<p>You can read more about these data by typing <code>?uitdewilligen</code> in R.</p>
<p>Suppose we wanted to use parallel computing to genotype this individual. Also suppose we only wanted the MAP (maximum a posteriori) estimated genotypes for each individual and each gene. Then we can use the following code.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nc &lt;-<span class="st"> </span><span class="dv">2</span> ## number of cores. 
        ## You should change this for your specific computing environment.
cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(nc)
doParallel<span class="op">::</span><span class="kw">registerDoParallel</span>(<span class="dt">cl =</span> cl)
<span class="kw">stopifnot</span>(foreach<span class="op">::</span><span class="kw">getDoParWorkers</span>() <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) ## make sure cluster is set up.
gene_est &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>ngenes, 
                    <span class="dt">.combine =</span> cbind, 
                    <span class="dt">.export =</span> <span class="st">&quot;flexdog&quot;</span>) <span class="op">%dopar%</span><span class="st"> </span>{
                      ## fit flexdog
                      fout &lt;-<span class="st"> </span><span class="kw">flexdog</span>(<span class="dt">refvec  =</span> uitdewilligen<span class="op">$</span>refmat[, i], 
                                      <span class="dt">sizevec =</span> uitdewilligen<span class="op">$</span>sizemat[, i], 
                                      <span class="dt">ploidy  =</span> uitdewilligen<span class="op">$</span>ploidy, 
                                      <span class="dt">model   =</span> <span class="st">&quot;norm&quot;</span>, 
                                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
                      fout<span class="op">$</span>geno
                      }
<span class="kw">stopCluster</span>(cl)</code></pre></div>
<p>Note that I used <code>model = &quot;norm&quot;</code> because we have a very small number of samples.</p>
<p>However, this doesn’t return any sort of diagnostic statistics for each SNP. So suppose we now want to save all of the output of each SNP and not just the MAP genotype, then we can use the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nc &lt;-<span class="st"> </span><span class="dv">2</span> ## number of cores. 
        ## You should change this for your specific computing environment.
cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(nc)
doParallel<span class="op">::</span><span class="kw">registerDoParallel</span>(<span class="dt">cl =</span> cl)
<span class="kw">stopifnot</span>(foreach<span class="op">::</span><span class="kw">getDoParWorkers</span>() <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) ## make sure cluster is set up.
glist &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>ngenes, 
                 <span class="dt">.export =</span> <span class="st">&quot;flexdog&quot;</span>) <span class="op">%dopar%</span><span class="st"> </span>{
                   ## fit flexdog
                   fout &lt;-<span class="st"> </span><span class="kw">flexdog</span>(<span class="dt">refvec  =</span> uitdewilligen<span class="op">$</span>refmat[, i], 
                                   <span class="dt">sizevec =</span> uitdewilligen<span class="op">$</span>sizemat[, i], 
                                   <span class="dt">ploidy  =</span> uitdewilligen<span class="op">$</span>ploidy, 
                                   <span class="dt">model   =</span> <span class="st">&quot;norm&quot;</span>, 
                                   <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
                   fout
                   }
<span class="kw">stopCluster</span>(cl)</code></pre></div>
<p>Returning the diagnostics for each SNP is pretty easy. For example, below we extract the variable <code>prop_mis</code> from the list <code>glist</code> and make a histogram of these values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
prop_mis &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="dt">X =</span> glist, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x<span class="op">$</span>prop_mis)
<span class="kw">qplot</span>(prop_mis, 
      <span class="dt">xlab =</span> <span class="st">&quot;Posterior Proportion Mis-genotyped&quot;</span>, 
      <span class="dt">bins =</span> <span class="dv">30</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAbAAAAFQCAMAAADp1NuaAAAA51BMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6ADo6AGY6OpA6kNtNTU1NTW5NTY5NbqtNjshZWVlmAABmADpmAGZmOpBmZmZmkJBmtrZmtttmtv9uTU1uTW5uTY5ubo5ubqtuq8huq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQOjqQOmaQkDqQkGaQtpCQ2/+rbk2rbm6rjk2ryKur5OSr5P+2ZgC2Zjq225C22/+2/9u2///Ijk3I///bkDrbtmbb/7bb///kq27k///r6+v/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T///+BR72BAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAK1UlEQVR4nO2dC38axxVHkU1kFRw1apO2cloXNwlq64iklfqQTWvZiILEfv/P05nZN7AP2EG7f/ncX2KhZebuYQ4zexev2V5ASEWvbQBit0CYWCBMLBAmFggTC4SJxa7CviBain2Fbd/8qX6GHZp2IG0HEMKmCJNBQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBNDQJgYAsLEEDwL+yqNXfbvvS3CtgfCWkqLMBkEhIkhIEwMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkhIEwMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkh1Be2mgxfXgXB/Zvh6cd4G8JaSltH2LtxcHf6cTUZB7e/jLchrKW0NYTdf/c+/rH4/ftoI8JaSltD2OL1z3ZJXLz+GNz/6SqI/n3zp7XICFt/ivAYdYS9GltbZlWMhGVFJ8EMe5y0tWaYM5XOsGy/JBD2OGnrHMN+cKY4hnUibc0q0Uyv1WREldh+2jrCzAnYl+85D+tGWj7pkEFAmBgCwsQQECaGgDAxBISJISBMDAFhYggIE0NAmBgCwsQQECaGgDAxBISJISBMDAFhYggIE0NAmBgCwsQQECaGgDAxBISJISBMDAFhYggIE0NAmBgCwsQQECaGgDAxBISJISBMDAFhYggIE0NoKGz96wgywtr4NoTPJphhMggsiWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBND2EXYajLmLrPdSFtP2O1w7KRxH+fW09YStvjD9+OAO6V3Im0dYauf/j5xN0u3d7i3fWysfx1BRlgb34bw2UQdYbcjuxzencbCsqKTYIY9TtoawszUWuVmWLZfEgh7nLQ1hN0ObYw4hnUibf2yfjUZUSW2n5bzMBkEPukQQ0CYGALCxBCKhS3Pzu2P2bOb4t4IayktwmQQioRNe3H0S3ojrKW0JTOsNBDWUlqKDhmEEmHzY7ckcgzrFEKxsIeLsqNXkOuXBMIeJy3HMBmEshmGsA4ilBzDSs/Agly/JBD2OGm3L4k9io7uIVDWiyEgTAyBJVEMoWqGLX91WdIbYS2lLVkSZ88/FPdGWEtpy4SxJHYKoVLYNTOsUwiVRccRx7BOIVDWiyEgTAyhTJi7TGBQ1hthLaXdKmxq68PlWZkxhLWUlqumZBAQJobAkiiGQNEhhkBZL4aAMDGEEmEPF4Oqa90Q1lLarcKuratyY3ztQ0tBWS+DwHmYGALnYWIIZVXijPOw7iFQ1oshIEwMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkhIEwMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkhIEwMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkhIEwMAWFiCAgTQ0CYGALCxBAQJoZQX9ji1XDIXWY7kraGMHt79MW3V/bWwNzHufW0NYTdWUvvxtwpvRNpax7DzCxbvP7oJpvpY2P96wgywtr4NoTPJuoJs7e1vzuNhWVFJ8EMe5y0tYTdvxmZ0uM1wjqQtl6VaGrEgGNYJ9LWEBb6cssiVWLraWsIux3aGHMe1om0fNIhg4AwMQSEiSEgTAwBYWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIICBNDQJgYAsLEEBAmhoAwMQSEiSEgTAwBYWIIhxP2VS15WqPVAQSEiSEgTAwBYWIIDYWtfx1BgbA2vhnhSQczTAaBJVEMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkhIEwMAWFiCAgTQ0CYGALCxBAQJoaAMDEEhIkhIEwMAWFiCAgTQ0CYGALCxBBaFbaj3fZHqwMICEMYwg6fFmH+20oLK9CBsH3TIsx/W4R1YLQ6gIAwhPmM6pe1vTXCtgfCWkqLMP9tEYawtOkOwmrftNS3sJ02fypovW++JG/1AK01rQYpSru9467CVpNxzdsCe4mC1JWbERZH/Rtve4mC1JWbERZHemv7L2y0/a/pP9eoL+zuNBaWFV3nDdO4aQfSdgBh/xmW7dd0/97bIiyOimPYnvv33hZhcawmo7Iqcc/9e2+LsCTKz8P23L/3tgjbHghrKS3CZBAQJoaAMDEEhIkhIEwMAWFiCA2FES3FnsKKPPpJQ9rKtAgTS4swsbSehBGPFQgTC4SJBcLEAmFi0URY+jeb0aPcX3X6S3s3HH75vqLTHmlXk+HLq4pOe6RdvPJLG4QXaJi0w3HQSFh6hWn0KH/Jqbe0lvcAaYN3Y3dFmOe092/MD49p7fvVvAHsZVCLb68aCUuvzoke5S/X8ZbWbjxAWvujeWykzV9k1jxt8O7l38yDO2vPvMWaCEvRokd+WDfS2o3NZ9gW2p89LIkHH4Qgfbe6DQ2EpVeYRo/yl5x6S2vX7+Yju5F28WrsxsVzWrckNsbNDWUkzF6+JjPDDpD2ULSmOvjjTweYYfdvrC+VY5hbvv2mvf/B78EmQ9v88JgbyqhKDEegUZU4SgqkUVgljryUc2tp/ay0G2nte6D5kriR1pL6HIQgFBb78nAeZtMd4Dwsk/Z26O+EKU/r7YQpk9acNXodhFCYGYOhOxHjkw6xQJhYIEwsECYWCBMLhIkFwsQCYWLRnrCHi56No8vNp/73r/zv8xebjUq6b024LYdNffzsxvxYnj3/UNCiNGtZzFxm39GmsIH9Md18WbVGrrj7ZpQknB875fPj5x+q89TO6uKJCluena8/s4uwLd03o1TYNzbP9BuEVUVmxO3y1g/s6JlF7tz+OXDbzCuen/y19+yfZmiiNu73m1z3cFP8/Iu3vZ4d/Ex7G4N5JseLH80uQtHz499+/SF4+PPbcEkMCSLC5Vnv6MeTmwQl7hamcZjXFnvaz+3V4tmub5+msGs71v3A/u/esvPj83BszVBMzSgeuxG+TNrY3/Pd3ab0+TRf0t6pyOYwYxutpfPj3/3FuPjFf5ywmMDtYXk2MP+H6UKUsFsW186ih4vz7F5t07jrAYat9aLDvCq3dpg/5ifhK4xHwk0fO3pmQ9ImGs1Md7cp9/x6+0jYxrbA2ZmeB7PBLBR2khlj19wIKkQxGez6YDrlnjLb4q4HGLbWZ5iJmV1M7Mu/jlZG83DaS1cyNxRRm+TAkXZ3m3LP2yHL/h4LW98WOGGzfnB9HgqLCVxMXfOTm1KUad/+F+81bhp3PcCwdUqYW/vDt+40KgIeQdjy6//++mYWl/WOYJod9VKU+cm/L85TYVHTpy7MVtZxTWXXHrskRidYyShFbQqFxc+7xenkJtc+Fra+LQgPWP942w9m6XlYXHjGS18pysPFb06iVTnea6brAYatC8Liw7V7hWYYwsLRvEnN649HKVeY5LtHA1dQdETv/Y0cGWFTU5SGwmIClzctOrajhGKnru5M9ho2XZ71n2LRMUgfhoeOWfTRxXWv77YdXaaDm5Tk24UF2bK+n/3dPmkSrufICHMzOpxhs+yHJ3FtXoDiMKPJld3r0eXTLOsPEzt/vlQRs6oTanek8r3XkkBYYdgDkjuxKo3pwO9eqwJhxWFrxQpf0WeQCCOKAmFigTCxQJhYIEwsECYWCBOL/wN0qeGCmGaOlQAAAABJRU5ErkJggg==" /><!-- --></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-gerard2018harnessing">
<p>Gerard, David, Luis Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Harnessing Empirical Bayes and Mendelian Segregation for Genotyping Autopolyploids from Messy Sequencing Data.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. doi:<a href="https://doi.org/10.1101/281550">10.1101/281550</a>.</p>
</div>
<div id="ref-uitdewilligen2013next">
<p>Uitdewilligen, Anne-Marie A. AND D’hoop, Jan G. A. M. L. AND Wolters. 2013. “A Next-Generation Sequencing Method for Genotyping-by-Sequencing of Highly Heterozygous Autotetraploid Potato.” <em>PLOS ONE</em> 8 (5). Public Library of Science: 1–14. doi:<a href="https://doi.org/10.1371/journal.pone.0062355">10.1371/journal.pone.0062355</a>.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
