---
title: "Explore Gauss-Hermite Quadrature"
author: "David Gerard"
date: "December 5, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Abstract

I explore how many points I need in my Gauss-Hermite quadrature to get a good approximation to the integral of the beta-binomial.

Conclusion: Quadrature doesn't work well unless we have a large number of abscissas.

# Set up function

```{r}
library(updog)
library(SuppDists)
library(gaussquad)

qbetabinom <- function(p, size, alpha, beta) {
  qghyper(p = p, a = -alpha, k = size, N = -alpha - beta)
}

qbetabinom_mu_rho <- function(p, size, mu, rho) {
  alpha <- mu * (1 - rho) / rho
  beta  <- (1 - mu) * (1 - rho) / rho
  qbetabinom(p = p, size = size, alpha = alpha, beta = beta)
}

dbetabinom_hyper <- function(x, size, alpha, beta, log = FALSE) {
  dghyper(x = x, a = -alpha, k = size, N = -alpha - beta, log = log)
}

dbetabinom_hyper_mu_rho <- function(x, size, mu, rho, log = FALSE) {
  alpha <- mu * (1 - rho) / rho
  beta  <- (1 - mu) * (1 - rho) / rho
  dbetabinom_hyper(x = x, size = size, alpha = alpha, beta = beta, log = log) 
}

assertthat::are_equal(dbetabinom_hyper(2, 8, 2, 3), updog::dbetabinom(2, 8, 2, 3))

#' @param prob The allele dosage
#' @param eps The sequencing error rate.
eta_fun <- function(prob, eps) {
  prob * (1 - eps) + eps * (1 - prob)
}

#' @param eta probability after sequencing error rate
#' @param h The bias parameter
xi_fun <- function(eta, h) {
  eta / (h * (1 - eta) + eta)
}

#' bb function of z that we are trying to integrate
#' @param x the number of A allele reads
#' @param n The number of total reads
#' @param z The latent variable we are integrating over
#' @param eps The sequencing error rate
#' @param h The bias parameter
#' @param tau The od parameter of the reads
#' @param alpha The allele frequency
#' @param rho The inbreeding coefficient of individual i
#' @param ploidy The ploidy of the species.
obj <- function(x, n, z, eps, h, tau, alpha, rho, ploidy) {
  y <- qbetabinom_mu_rho(p = pnorm(z), size = ploidy, mu = alpha, rho = rho)
  eta <- eta_fun(prob = y / ploidy, eps = eps)  
  xi <- xi_fun(eta = eta, h = h)  
  dbetabinom_hyper_mu_rho(x = x, size = n, mu = xi, rho = tau, log = TRUE)
}


#' Guassian quadrature of \code{\link{obj}}.
#' @inheritParams obj
#' @param mu The mean of z
#' @param sigma2 The variance of z
#' @param npoints The number of abscissas.
#' @param grules A list of guass-hermite rules
equad <- function(x, n, eps, h, tau, alpha, rho, ploidy, mu, sigma2, npoints, grules) {
  
  zvec <- sqrt(2 * sigma2) * grules[[npoints]]$x + mu
  wvec <- grules[[npoints]]$w
  
  obj_val <- obj(x = x, n = n, z = zvec, eps = eps, h = h, tau = tau, alpha = alpha, rho = rho, ploidy = ploidy)
  
  sum(wvec * obj_val) / sqrt(pi)
}
```

# Input some reasonable parameter values
```{r}
x      <- 18
n      <- 100
z      <- -3
h      <- 1
tau    <- 0.01
alpha  <- 1/6
rho    <- 0.01
ploidy <- 6
sigma2 <- 1
mu     <- 0
eps    <- 0.01
```

# See how many we need for accurate integration

```{r}
nmax <- 100
grules <- hermite.h.quadrature.rules(n = nmax)
int_vec <- rep(NA, length = nmax - 1)
for (index in 1:(nmax - 1)) {
  npoints <- index + 1
  int_vec[index] <- equad(x = x, n = n, eps = eps, h = h, tau = tau, alpha = alpha, rho = rho, ploidy = ploidy, mu = mu, sigma2 = sigma2, npoints = npoints, grules = grules)
}

library(tidyverse)
integral_df <- data_frame(npoints = 2:nmax, int = int_vec)
```

Compare to stochastic approximation
```{r}
zvec <- rnorm(n = 1000, mean = mu, sd = sqrt(sigma2))
objout <- obj(x = x, n = n, z = zvec, eps = eps, h = h, tau = tau, alpha = alpha, rho = rho, ploidy = ploidy)
mean_obj <- mean(objout)
lower <- mean_obj - 2 * sd(objout) / sqrt(length(objout))
upper <- mean_obj + 2 * sd(objout) / sqrt(length(objout))
```

Plot the results
```{r}
ggplot(data = integral_df, mapping = aes(x = npoints, y = int_vec)) +
  geom_point() +
  theme_bw() +
  geom_hline(yintercept = mean_obj) +
  geom_hline(yintercept = lower, color = "blue", lty = 2, alpha = 1/2) +
  geom_hline(yintercept = upper, color = "blue", lty = 2, alpha = 1/2)
```

Weird pattern. What about the average of these guass quadrature rules?
```{r}
integral_df$cummean <- cumsum(integral_df$int) / 1:nrow(integral_df)
ggplot(data = integral_df, mapping = aes(x = npoints, y = cummean)) +
  geom_point() +
  theme_bw() +
  geom_hline(yintercept = mean_obj) +
  geom_hline(yintercept = lower, color = "blue", lty = 2, alpha = 1/2) +
  geom_hline(yintercept = upper, color = "blue", lty = 2, alpha = 1/2)
```

What about the mean of some of the later ones
```{r}
mean(integral_df$int[25:30])
```

