<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Oracle Calculations • updog</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Oracle Calculations">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multidog.html">Genotyping Many SNP's with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Oracle Calculations</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      
      <div class="hidden name"><code>oracle_calculations.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>We provide some example usage of the oracle calculations available in <code>updog</code>. These are particularly useful for read-depth determination. These calculations are described in detail in <span class="citation">Gerard et al. (2018)</span>.</p>
</div>
<div id="controlling-misclassification-error" class="section level1">
<h1 class="hasAnchor">
<a href="#controlling-misclassification-error" class="anchor"></a>Controlling Misclassification Error</h1>
<p>Suppose we have a sample of tetraploid individuals derived from an S1 cross (a single generation of selfing). Using domain expertise (either from previous studies or a pilot analysis), we’ve determined that our sequencing technology will produce relatively clean data. That is, the sequencing error rate will not be too large (say, ~0.001), the bias will be moderate (say, ~0.7 at the most extreme), and the majority of SNPs will have reasonable levels of overdispersion (say, less than 0.01). We want to know how deep we need to sequence.</p>
<p>Using <code>oracle_mis</code>, we can see how deep we need to sequence under the worst-case scenario we want to control (sequencing error rate = 0.001, bias = 0.7, overdispersion = 0.01) in order to obtain a misclassification error rate of at most, say, 0.05.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bias   &lt;-<span class="st"> </span><span class="fl">0.7</span>
od     &lt;-<span class="st"> </span><span class="fl">0.01</span>
seq    &lt;-<span class="st"> </span><span class="fl">0.001</span>
maxerr &lt;-<span class="st"> </span><span class="fl">0.05</span></code></pre></div>
<p>Before we do this, we also need the distribution of the offspring genotypes. We can get this distribution assuming various parental genotypes using the <code>get_q_array</code> function. Typically, error rates will be larger when the allele-frequency is closer to 0.5. So we’ll start in the worst-case scenario of assuming that the parent has 2 copies of the reference allele.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(updog)
ploidy &lt;-<span class="st"> </span><span class="dv">4</span>
pgeno &lt;-<span class="st"> </span><span class="dv">2</span>
gene_dist &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_q_array.html">get_q_array</a></span>(<span class="dt">ploidy =</span> ploidy)[pgeno <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, pgeno <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, ]</code></pre></div>
<p>This is what the genotype distribution for the offspring looks like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)
distdf &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">x =</span> <span class="dv">0</span><span class="op">:</span>ploidy, <span class="dt">y =</span> <span class="dv">0</span>, <span class="dt">yend =</span> gene_dist)
<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(distdf, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y, <span class="dt">xend =</span> x, <span class="dt">yend =</span> yend)) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span>(<span class="dt">lineend =</span> <span class="st">"round"</span>, <span class="dt">lwd =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Allele Dosage"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"Probability"</span>)</code></pre></div>
<p><img src="oracle_calculations_files/figure-html/unnamed-chunk-3-1.png" width="432"></p>
<p>Now, we are ready to iterate through read-depth’s until we reach one with an error rate less than 0.05.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">err    &lt;-<span class="st"> </span><span class="ot">Inf</span>
depth  &lt;-<span class="st"> </span><span class="dv">0</span>
<span class="cf">while</span>(err <span class="op">&gt;</span><span class="st"> </span>maxerr) {
  depth &lt;-<span class="st"> </span>depth <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
  err &lt;-<span class="st"> </span><span class="kw"><a href="../reference/oracle_mis.html">oracle_mis</a></span>(<span class="dt">n =</span> depth,
                    <span class="dt">ploidy =</span> ploidy,
                    <span class="dt">seq =</span> seq,
                    <span class="dt">bias =</span> bias,
                    <span class="dt">od =</span> od,
                    <span class="dt">dist =</span> gene_dist)
}
depth
<span class="co">#&gt; [1] 90</span></code></pre></div>
<p>Looks like we need a depth of 90 in order to get a misclassification error rate under 0.05.</p>
<p>Note that <code>oracle_mis</code> returns the <strong>best misclassification error rate possible</strong> under these conditions (<code>ploidy</code> = 4, <code>bias</code> = 0.7, <code>seq</code> = 0.001, <code>od</code> = 0.01, and <code>pgeno</code> = 2). In your actual analysis, you will have a worse misclassification error rate than that returned by <code>oracle_mis</code>. However, if you have a lot of individuals in your sample, then this will act as a reasonable approximation to the error rate. In general though, you should sequence a little deeper than suggested by <code>oracle_mis</code>.</p>
</div>
<div id="visualizing-the-joint-distribution" class="section level1">
<h1 class="hasAnchor">
<a href="#visualizing-the-joint-distribution" class="anchor"></a>Visualizing the Joint Distribution</h1>
<p>Suppose we only have a budget to sequence to a depth of 30. Then what errors can we expect? We can use <code>oracle_joint</code> and <code>oracle_plot</code> to visualize the errors we can expect.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">depth &lt;-<span class="st"> </span><span class="dv">30</span>
jd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/oracle_joint.html">oracle_joint</a></span>(<span class="dt">n =</span> depth,
                   <span class="dt">ploidy =</span> ploidy,
                   <span class="dt">seq =</span> seq,
                   <span class="dt">bias =</span> bias,
                   <span class="dt">od =</span> od,
                   <span class="dt">dist =</span> gene_dist)
<span class="kw"><a href="../reference/oracle_plot.html">oracle_plot</a></span>(jd)</code></pre></div>
<p><img src="oracle_calculations_files/figure-html/unnamed-chunk-5-1.png" width="432"></p>
<p>Most of the errors will be mistakes between genotypes 2/3 and mistakes between genotypes 1/2.</p>
<p>Even though the misclassification error rate is pretty high (0.14), the correlation of the oracle estimator with the true genotype is pretty reasonable (0.89). You can obtain this using the <code>oracle_cor</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ocorr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/oracle_cor.html">oracle_cor</a></span>(<span class="dt">n =</span> depth,
                    <span class="dt">ploidy =</span> ploidy,
                    <span class="dt">seq =</span> seq,
                    <span class="dt">bias =</span> bias,
                    <span class="dt">od =</span> od,
                    <span class="dt">dist =</span> gene_dist)
ocorr
<span class="co">#&gt; [1] 0.8935101</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-gerard2018genotyping">
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. doi:<a href="https://doi.org/10.1534/genetics.118.301468">10.1534/genetics.118.301468</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li><a href="#controlling-misclassification-error">Controlling Misclassification Error</a></li>
      <li><a href="#visualizing-the-joint-distribution">Visualizing the Joint Distribution</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
