<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Parallelized Genotyping using Updog • updog</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Parallelized Genotyping using Updog">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Parallelized Genotyping using Updog</h1>
                        <h4 class="author">David Gerard</h4>
            
            <h4 class="date">2019-09-09</h4>
      
      
      <div class="hidden name"><code>parallel_computing.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>We provide two examples of parallelizing <code>flexdog</code> to genotype many genes.</p>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>Before we begin, we’ll need to load the <code>parallel</code>, <code>foreach</code>, and <code>doParallel</code> packages. There are other ways to perform parallel computation in R, but we find these to be pretty handy and easy to use.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(parallel)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(foreach)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(doParallel)</a></code></pre></div>
<p>Of course, we’ll need to load the <code>updog</code> package to actually do the genotyping.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(updog)</a></code></pre></div>
<p>We’ll work with the <code>uitedewilligen</code> dataset, which is a subset of the data from <span class="citation">Uitdewilligen (2013)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"uitdewilligen"</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">ngenes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(uitdewilligen<span class="op">$</span>refmat)</a></code></pre></div>
<p>You can read more about these data by typing <code><a href="../reference/uitdewilligen.html">?uitdewilligen</a></code> in R.</p>
<p>Suppose we wanted to use parallel computing to genotype this individual. Also suppose we only wanted the MAP (maximum a posteriori) estimated genotypes for each individual and each gene. Then we can use the following code.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">nc &lt;-<span class="st"> </span><span class="dv">2</span> <span class="co">## number of cores. </span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">        <span class="co">## You should change this for your specific computing environment.</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/parallel/makeCluster.html">makeCluster</a></span>(nc)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">doParallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html">registerDoParallel</a></span>(<span class="dt">cl =</span> cl)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span>(foreach<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/foreach/man/getDoParWorkers.html">getDoParWorkers</a></span>() <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="co">## make sure cluster is set up.</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">gene_est &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/foreach/man/foreach.html">foreach</a></span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>ngenes, </a>
<a class="sourceLine" id="cb4-7" data-line-number="7">                    <span class="dt">.combine =</span> cbind, </a>
<a class="sourceLine" id="cb4-8" data-line-number="8">                    <span class="dt">.export =</span> <span class="st">"flexdog"</span>) <span class="op">%dopar%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">                      <span class="co">## fit flexdog</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">                      fout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> uitdewilligen<span class="op">$</span>refmat[, i], </a>
<a class="sourceLine" id="cb4-11" data-line-number="11">                                      <span class="dt">sizevec =</span> uitdewilligen<span class="op">$</span>sizemat[, i], </a>
<a class="sourceLine" id="cb4-12" data-line-number="12">                                      <span class="dt">ploidy  =</span> uitdewilligen<span class="op">$</span>ploidy, </a>
<a class="sourceLine" id="cb4-13" data-line-number="13">                                      <span class="dt">model   =</span> <span class="st">"norm"</span>, </a>
<a class="sourceLine" id="cb4-14" data-line-number="14">                                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-15" data-line-number="15">                      fout<span class="op">$</span>geno</a>
<a class="sourceLine" id="cb4-16" data-line-number="16">                      }</a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="kw"><a href="https://rdrr.io/r/parallel/makeCluster.html">stopCluster</a></span>(cl)</a></code></pre></div>
<p>Note that I used <code>model = "norm"</code> <span class="citation">(Gerard and Ferrão 2019)</span> because we have a very small number of samples.</p>
<p>However, this doesn’t return any sort of diagnostic statistics for each SNP. So suppose we now want to save all of the output of each SNP and not just the MAP genotype, then we can use the following code:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">nc &lt;-<span class="st"> </span><span class="dv">2</span> <span class="co">## number of cores. </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">        <span class="co">## You should change this for your specific computing environment.</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/parallel/makeCluster.html">makeCluster</a></span>(nc)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">doParallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html">registerDoParallel</a></span>(<span class="dt">cl =</span> cl)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span>(foreach<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/foreach/man/getDoParWorkers.html">getDoParWorkers</a></span>() <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="co">## make sure cluster is set up.</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">glist &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/foreach/man/foreach.html">foreach</a></span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>ngenes, </a>
<a class="sourceLine" id="cb5-7" data-line-number="7">                 <span class="dt">.export =</span> <span class="st">"flexdog"</span>) <span class="op">%dopar%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">                   <span class="co">## fit flexdog</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">                   fout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> uitdewilligen<span class="op">$</span>refmat[, i], </a>
<a class="sourceLine" id="cb5-10" data-line-number="10">                                   <span class="dt">sizevec =</span> uitdewilligen<span class="op">$</span>sizemat[, i], </a>
<a class="sourceLine" id="cb5-11" data-line-number="11">                                   <span class="dt">ploidy  =</span> uitdewilligen<span class="op">$</span>ploidy, </a>
<a class="sourceLine" id="cb5-12" data-line-number="12">                                   <span class="dt">model   =</span> <span class="st">"norm"</span>, </a>
<a class="sourceLine" id="cb5-13" data-line-number="13">                                   <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-14" data-line-number="14">                   fout</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">                   }</a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="kw"><a href="https://rdrr.io/r/parallel/makeCluster.html">stopCluster</a></span>(cl)</a></code></pre></div>
<p>Returning the diagnostics for each SNP is pretty easy. For example, below we extract the variable <code>prop_mis</code> from the list <code>glist</code> and make a histogram of these values.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">prop_mis &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="dt">X =</span> glist, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x<span class="op">$</span>prop_mis)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw">qplot</span>(prop_mis, </a>
<a class="sourceLine" id="cb6-4" data-line-number="4">      <span class="dt">xlab =</span> <span class="st">"Posterior Proportion Mis-genotyped"</span>, </a>
<a class="sourceLine" id="cb6-5" data-line-number="5">      <span class="dt">bins =</span> <span class="dv">30</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="st">  </span><span class="kw">theme_bw</span>()</a></code></pre></div>
<p><img src="parallel_computing_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-gerard2018genotyping">
<p>Gerard, David, Luis Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468">https://doi.org/10.1534/genetics.118.301468</a>.</p>
</div>
<div id="ref-gerard2019priors">
<p>Gerard, David, and Luís Felipe Ventorim Ferrão. 2019. “Priors for Genotyping Polyploids.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/751784">https://doi.org/10.1101/751784</a>.</p>
</div>
<div id="ref-uitdewilligen2013next">
<p>Uitdewilligen, Anne-Marie A. AND D’hoop, Jan G. A. M. L. AND Wolters. 2013. “A Next-Generation Sequencing Method for Genotyping-by-Sequencing of Highly Heterozygous Autotetraploid Potato.” <em>PLOS ONE</em> 8 (5). Public Library of Science: 1–14. <a href="https://doi.org/10.1371/journal.pone.0062355">https://doi.org/10.1371/journal.pone.0062355</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li><a href="#analysis">Analysis</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
