<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulate Next-Generation Sequencing Data • updog</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Next-Generation Sequencing Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Simulate Next-Generation Sequencing Data</h1>
                        <h4 class="author">David Gerard</h4>
            
            <h4 class="date">2018-11-07</h4>
      
      
      <div class="hidden name"><code>simulate_ngs.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>We demonstrate how to simulate NGS data under various genotype distributions, then fit these data using <code>flexdog</code>.</p>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>Let’s suppose that we have 100 hexaploid individuals, with varying levels of read-depth.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(updog)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">nind    &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">ploidy  &lt;-<span class="st"> </span><span class="dv">6</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">sizevec &lt;-<span class="st"> </span><span class="kw">round</span>(stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dt">n   =</span> nind, </a>
<a class="sourceLine" id="cb1-6" data-line-number="6">                              <span class="dt">min =</span> <span class="dv">50</span>, </a>
<a class="sourceLine" id="cb1-7" data-line-number="7">                              <span class="dt">max =</span> <span class="dv">200</span>))</a></code></pre></div>
<p>We can simulate their read-counts under various genotype distributions, allele biases, overdispersions, and sequencing error rates using the <code>rgeno</code> and <code>rflexdog</code> functions.</p>
<div id="f1-population" class="section level2">
<h2 class="hasAnchor">
<a href="#f1-population" class="anchor"></a>F1 Population</h2>
<p>Suppose these individuals are all siblings where the first parent has 4 copies of the reference allele and the second parent has 5 copies of the reference allele. Then the following code, using <code>rgeno</code>, will simulate the individuals’ genotypes.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">true_geno &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rgeno.html">rgeno</a></span>(<span class="dt">n      =</span> nind, </a>
<a class="sourceLine" id="cb2-2" data-line-number="2">                   <span class="dt">ploidy =</span> ploidy,</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">                   <span class="dt">model  =</span> <span class="st">"f1"</span>, </a>
<a class="sourceLine" id="cb2-4" data-line-number="4">                   <span class="dt">p1geno =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">                   <span class="dt">p2geno =</span> <span class="dv">5</span>)</a></code></pre></div>
<p>Once we have their genotypes, we can simulate their read-counts using <code>rflexdog</code>. Let’s suppose that there is a moderate level of allelic bias (0.7) and a small level of overdispersion (0.005). Generally, in the real data that I’ve seen, the bias will range between 0.5 and 2 and the overdispersion will range between 0 and 0.02, with only a few extremely overdispersed SNPs above 0.02.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">refvec &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rflexdog.html">rflexdog</a></span>(<span class="dt">sizevec =</span> sizevec, </a>
<a class="sourceLine" id="cb3-2" data-line-number="2">                   <span class="dt">geno    =</span> true_geno, </a>
<a class="sourceLine" id="cb3-3" data-line-number="3">                   <span class="dt">ploidy  =</span> ploidy, </a>
<a class="sourceLine" id="cb3-4" data-line-number="4">                   <span class="dt">seq     =</span> <span class="fl">0.001</span>, </a>
<a class="sourceLine" id="cb3-5" data-line-number="5">                   <span class="dt">bias    =</span> <span class="fl">0.7</span>, </a>
<a class="sourceLine" id="cb3-6" data-line-number="6">                   <span class="dt">od      =</span> <span class="fl">0.005</span>)</a></code></pre></div>
<p>When we plot the data, it looks realistic</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="../reference/plot_geno.html">plot_geno</a></span>(<span class="dt">refvec  =</span> refvec, </a>
<a class="sourceLine" id="cb4-2" data-line-number="2">          <span class="dt">sizevec =</span> sizevec, </a>
<a class="sourceLine" id="cb4-3" data-line-number="3">          <span class="dt">ploidy  =</span> ploidy, </a>
<a class="sourceLine" id="cb4-4" data-line-number="4">          <span class="dt">bias    =</span> <span class="fl">0.7</span>, </a>
<a class="sourceLine" id="cb4-5" data-line-number="5">          <span class="dt">seq     =</span> <span class="fl">0.001</span>,</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">          <span class="dt">geno    =</span> true_geno)</a></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
<p>We can test <code>flexdog</code> on these data</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">fout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> refvec, </a>
<a class="sourceLine" id="cb5-2" data-line-number="2">                <span class="dt">sizevec =</span> sizevec, </a>
<a class="sourceLine" id="cb5-3" data-line-number="3">                <span class="dt">ploidy  =</span> ploidy, </a>
<a class="sourceLine" id="cb5-4" data-line-number="4">                <span class="dt">model   =</span> <span class="st">"f1"</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt;          Fit: 1 of 5 </span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; Initial Bias: 0.3678794 </span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; Log-Likelihood: -369.5682 </span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt; Keeping new fit.</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt;          Fit: 2 of 5 </span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt; Initial Bias: 0.6065307 </span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt; Log-Likelihood: -369.5683 </span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt; Keeping old fit.</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt;          Fit: 3 of 5 </span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; Initial Bias: 1 </span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; Log-Likelihood: -369.5682 </span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; Keeping new fit.</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt;          Fit: 4 of 5 </span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt; Initial Bias: 1.648721 </span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; Log-Likelihood: -379.0568 </span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt; Keeping old fit.</span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25"><span class="co">#&gt;          Fit: 5 of 5 </span></a>
<a class="sourceLine" id="cb5-26" data-line-number="26"><span class="co">#&gt; Initial Bias: 2.718282 </span></a>
<a class="sourceLine" id="cb5-27" data-line-number="27"><span class="co">#&gt; Log-Likelihood: -401.265 </span></a>
<a class="sourceLine" id="cb5-28" data-line-number="28"><span class="co">#&gt; Keeping old fit.</span></a>
<a class="sourceLine" id="cb5-29" data-line-number="29"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-30" data-line-number="30"><span class="co">#&gt; Done!</span></a></code></pre></div>
<p><code>flexdog</code> gives us reasonable genotyping, and it accurately estimates the proportion of individuals mis-genotyped.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">plot</span>(fout)</a></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">## Estimated proportion misgenotyped</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">fout<span class="op">$</span>prop_mis</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#&gt; [1] 0.0703387</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">## Actual proportion misgenotyped</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="kw">mean</span>(fout<span class="op">$</span>geno <span class="op">!=</span><span class="st"> </span>true_geno)</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt; [1] 0.05</span></a></code></pre></div>
</div>
<div id="hwe-population" class="section level2">
<h2 class="hasAnchor">
<a href="#hwe-population" class="anchor"></a>HWE Population</h2>
<p>Now run the same simulations assuming the individuals are in Hardy-Weinberg population with an allele frequency of 0.75.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">true_geno &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rgeno.html">rgeno</a></span>(<span class="dt">n      =</span> nind,</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">                   <span class="dt">ploidy =</span> ploidy,</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">                   <span class="dt">model  =</span> <span class="st">"hw"</span>, </a>
<a class="sourceLine" id="cb8-4" data-line-number="4">                   <span class="dt">allele_freq =</span> <span class="fl">0.75</span>)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5">refvec &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rflexdog.html">rflexdog</a></span>(<span class="dt">sizevec =</span> sizevec, </a>
<a class="sourceLine" id="cb8-6" data-line-number="6">                   <span class="dt">geno    =</span> true_geno, </a>
<a class="sourceLine" id="cb8-7" data-line-number="7">                   <span class="dt">ploidy  =</span> ploidy, </a>
<a class="sourceLine" id="cb8-8" data-line-number="8">                   <span class="dt">seq     =</span> <span class="fl">0.001</span>, </a>
<a class="sourceLine" id="cb8-9" data-line-number="9">                   <span class="dt">bias    =</span> <span class="fl">0.7</span>, </a>
<a class="sourceLine" id="cb8-10" data-line-number="10">                   <span class="dt">od      =</span> <span class="fl">0.005</span>)</a>
<a class="sourceLine" id="cb8-11" data-line-number="11">fout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> refvec, </a>
<a class="sourceLine" id="cb8-12" data-line-number="12">                <span class="dt">sizevec =</span> sizevec, </a>
<a class="sourceLine" id="cb8-13" data-line-number="13">                <span class="dt">ploidy  =</span> ploidy, </a>
<a class="sourceLine" id="cb8-14" data-line-number="14">                <span class="dt">model   =</span> <span class="st">"hw"</span>)</a>
<a class="sourceLine" id="cb8-15" data-line-number="15"><span class="co">#&gt;          Fit: 1 of 5 </span></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="co">#&gt; Initial Bias: 0.3678794 </span></a>
<a class="sourceLine" id="cb8-17" data-line-number="17"><span class="co">#&gt; Log-Likelihood: -383.0611 </span></a>
<a class="sourceLine" id="cb8-18" data-line-number="18"><span class="co">#&gt; Keeping new fit.</span></a>
<a class="sourceLine" id="cb8-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20"><span class="co">#&gt;          Fit: 2 of 5 </span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21"><span class="co">#&gt; Initial Bias: 0.6065307 </span></a>
<a class="sourceLine" id="cb8-22" data-line-number="22"><span class="co">#&gt; Log-Likelihood: -383.0611 </span></a>
<a class="sourceLine" id="cb8-23" data-line-number="23"><span class="co">#&gt; Keeping new fit.</span></a>
<a class="sourceLine" id="cb8-24" data-line-number="24"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-25" data-line-number="25"><span class="co">#&gt;          Fit: 3 of 5 </span></a>
<a class="sourceLine" id="cb8-26" data-line-number="26"><span class="co">#&gt; Initial Bias: 1 </span></a>
<a class="sourceLine" id="cb8-27" data-line-number="27"><span class="co">#&gt; Log-Likelihood: -383.061 </span></a>
<a class="sourceLine" id="cb8-28" data-line-number="28"><span class="co">#&gt; Keeping new fit.</span></a>
<a class="sourceLine" id="cb8-29" data-line-number="29"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-30" data-line-number="30"><span class="co">#&gt;          Fit: 4 of 5 </span></a>
<a class="sourceLine" id="cb8-31" data-line-number="31"><span class="co">#&gt; Initial Bias: 1.648721 </span></a>
<a class="sourceLine" id="cb8-32" data-line-number="32"><span class="co">#&gt; Log-Likelihood: -383.061 </span></a>
<a class="sourceLine" id="cb8-33" data-line-number="33"><span class="co">#&gt; Keeping old fit.</span></a>
<a class="sourceLine" id="cb8-34" data-line-number="34"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-35" data-line-number="35"><span class="co">#&gt;          Fit: 5 of 5 </span></a>
<a class="sourceLine" id="cb8-36" data-line-number="36"><span class="co">#&gt; Initial Bias: 2.718282 </span></a>
<a class="sourceLine" id="cb8-37" data-line-number="37"><span class="co">#&gt; Log-Likelihood: -383.061 </span></a>
<a class="sourceLine" id="cb8-38" data-line-number="38"><span class="co">#&gt; Keeping old fit.</span></a>
<a class="sourceLine" id="cb8-39" data-line-number="39"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-40" data-line-number="40"><span class="co">#&gt; Done!</span></a>
<a class="sourceLine" id="cb8-41" data-line-number="41"><span class="kw">plot</span>(fout)</a></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-7-1.png" width="432"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co">## Estimated proportion misgenotyped</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">fout<span class="op">$</span>prop_mis</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; [1] 0.09165815</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">## Actual proportion misgenotyped</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="kw">mean</span>(fout<span class="op">$</span>geno <span class="op">!=</span><span class="st"> </span>true_geno)</a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">#&gt; [1] 0.07</span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co">## Estimated allele frequency close to true allele frequency</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11">fout<span class="op">$</span>par<span class="op">$</span>alpha</a>
<a class="sourceLine" id="cb9-12" data-line-number="12"><span class="co">#&gt; [1] 0.7498848</span></a></code></pre></div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-gerard2018genotyping">
<p>Gerard, David, Luis Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468" class="uri">https://doi.org/10.1534/genetics.118.301468</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li>
<a href="#analysis">Analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#f1-population">F1 Population</a></li>
      <li><a href="#hwe-population">HWE Population</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
