<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulate Next-Generation Sequencing Data • updog</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Next-Generation Sequencing Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multidog.html">Genotyping Many SNP's with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Simulate Next-Generation Sequencing Data</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      
      <div class="hidden name"><code>simulate_ngs.Rmd</code></div>

    </div>

    
    
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>We demonstrate how to simulate NGS data under various genotype distributions, then fit these data using <code>flexdog</code>. The genotyping methods are described in <span class="citation">Gerard et al. (2018)</span>.</p>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<p>Let’s suppose that we have 100 hexaploid individuals, with varying levels of read-depth.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">1</span>)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(updog)
nind    &lt;-<span class="st"> </span><span class="dv">100</span>
ploidy  &lt;-<span class="st"> </span><span class="dv">6</span>
sizevec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(stats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span>(<span class="dt">n   =</span> nind, 
                              <span class="dt">min =</span> <span class="dv">50</span>, 
                              <span class="dt">max =</span> <span class="dv">200</span>))</code></pre></div>
<p>We can simulate their read-counts under various genotype distributions, allele biases, overdispersions, and sequencing error rates using the <code>rgeno</code> and <code>rflexdog</code> functions.</p>
<div id="f1-population" class="section level2">
<h2 class="hasAnchor">
<a href="#f1-population" class="anchor"></a>F1 Population</h2>
<p>Suppose these individuals are all siblings where the first parent has 4 copies of the reference allele and the second parent has 5 copies of the reference allele. Then the following code, using <code>rgeno</code>, will simulate the individuals’ genotypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">true_geno &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rgeno.html">rgeno</a></span>(<span class="dt">n      =</span> nind, 
                   <span class="dt">ploidy =</span> ploidy,
                   <span class="dt">model  =</span> <span class="st">"f1"</span>, 
                   <span class="dt">p1geno =</span> <span class="dv">4</span>,
                   <span class="dt">p2geno =</span> <span class="dv">5</span>)</code></pre></div>
<p>Once we have their genotypes, we can simulate their read-counts using <code>rflexdog</code>. Let’s suppose that there is a moderate level of allelic bias (0.7) and a small level of overdispersion (0.005). Generally, in the real data that I’ve seen, the bias will range between 0.5 and 2 and the overdispersion will range between 0 and 0.02, with only a few extremely overdispersed SNPs above 0.02.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">refvec &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rflexdog.html">rflexdog</a></span>(<span class="dt">sizevec =</span> sizevec, 
                   <span class="dt">geno    =</span> true_geno, 
                   <span class="dt">ploidy  =</span> ploidy, 
                   <span class="dt">seq     =</span> <span class="fl">0.001</span>, 
                   <span class="dt">bias    =</span> <span class="fl">0.7</span>, 
                   <span class="dt">od      =</span> <span class="fl">0.005</span>)</code></pre></div>
<p>When we plot the data, it looks realistic</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_geno.html">plot_geno</a></span>(<span class="dt">refvec  =</span> refvec, 
          <span class="dt">sizevec =</span> sizevec, 
          <span class="dt">ploidy  =</span> ploidy, 
          <span class="dt">bias    =</span> <span class="fl">0.7</span>, 
          <span class="dt">seq     =</span> <span class="fl">0.001</span>,
          <span class="dt">geno    =</span> true_geno)</code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
<p>We can test <code>flexdog</code> on these data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> refvec, 
                <span class="dt">sizevec =</span> sizevec, 
                <span class="dt">ploidy  =</span> ploidy, 
                <span class="dt">model   =</span> <span class="st">"f1"</span>)
<span class="co">#&gt;          Fit: 1 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.3678794 </span>
<span class="co">#&gt; Log-Likelihood: -363.9369 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 2 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.6065307 </span>
<span class="co">#&gt; Log-Likelihood: -363.937 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 3 of 5 </span>
<span class="co">#&gt; Initial Bias: 1 </span>
<span class="co">#&gt; Log-Likelihood: -363.9369 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 4 of 5 </span>
<span class="co">#&gt; Initial Bias: 1.648721 </span>
<span class="co">#&gt; Log-Likelihood: -381.6123 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 5 of 5 </span>
<span class="co">#&gt; Initial Bias: 2.718282 </span>
<span class="co">#&gt; Log-Likelihood: -412.7397 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Done!</span></code></pre></div>
<p><code>flexdog</code> gives us reasonable genotyping, and it accurately estimates the proportion of individuals mis-genotyped.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(fout)</code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
## Estimated proportion misgenotyped
fout<span class="op">$</span>prop_mis
<span class="co">#&gt; [1] 0.07011089</span>

## Actual proportion misgenotyped
<span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(fout<span class="op">$</span>geno <span class="op">!=</span><span class="st"> </span>true_geno)
<span class="co">#&gt; [1] 0.05</span></code></pre></div>
</div>
<div id="hwe-population" class="section level2">
<h2 class="hasAnchor">
<a href="#hwe-population" class="anchor"></a>HWE Population</h2>
<p>Now run the same simulations assuming the individuals are in Hardy-Weinberg population with an allele frequency of 0.75.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">true_geno &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rgeno.html">rgeno</a></span>(<span class="dt">n      =</span> nind,
                   <span class="dt">ploidy =</span> ploidy,
                   <span class="dt">model  =</span> <span class="st">"hw"</span>, 
                   <span class="dt">allele_freq =</span> <span class="fl">0.75</span>)
refvec &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rflexdog.html">rflexdog</a></span>(<span class="dt">sizevec =</span> sizevec, 
                   <span class="dt">geno    =</span> true_geno, 
                   <span class="dt">ploidy  =</span> ploidy, 
                   <span class="dt">seq     =</span> <span class="fl">0.001</span>, 
                   <span class="dt">bias    =</span> <span class="fl">0.7</span>, 
                   <span class="dt">od      =</span> <span class="fl">0.005</span>)
fout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> refvec, 
                <span class="dt">sizevec =</span> sizevec, 
                <span class="dt">ploidy  =</span> ploidy, 
                <span class="dt">model   =</span> <span class="st">"hw"</span>)
<span class="co">#&gt;          Fit: 1 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.3678794 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 2 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.6065307 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 3 of 5 </span>
<span class="co">#&gt; Initial Bias: 1 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 4 of 5 </span>
<span class="co">#&gt; Initial Bias: 1.648721 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 5 of 5 </span>
<span class="co">#&gt; Initial Bias: 2.718282 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Done!</span>
<span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(fout)</code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-7-1.png" width="432"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
## Estimated proportion misgenotyped
fout<span class="op">$</span>prop_mis
<span class="co">#&gt; [1] 0.07625985</span>

## Actual proportion misgenotyped
<span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(fout<span class="op">$</span>geno <span class="op">!=</span><span class="st"> </span>true_geno)
<span class="co">#&gt; [1] 0.07</span>

## Estimated allele frequency close to true allele frequency
fout<span class="op">$</span>par<span class="op">$</span>alpha
<span class="co">#&gt; [1] 0.7473264</span></code></pre></div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-gerard2018genotyping">
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. doi:<a href="https://doi.org/10.1534/genetics.118.301468">10.1534/genetics.118.301468</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li>
<a href="#analysis">Analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#f1-population">F1 Population</a></li>
      <li><a href="#hwe-population">HWE Population</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
