<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fit <code><a href='flexdog.html'>flexdog</a></code> to multiple SNPs. — multidog • updog</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fit <code><a href='flexdog.html'>flexdog</a></code> to multiple SNPs. — multidog" />
<meta property="og:description" content="This is a convenience function that will run flexdog over many SNPs.
Support is provided for parallel computing through the doParallel package.
This function has not been extensively tested. Please report any bugs to
http://github.com/dcgerard/updog/issues." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/dcgerard/updog/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit <code><a href='flexdog.html'>flexdog</a></code> to multiple SNPs.</h1>
    <small class="dont-index">Source: <a href='http://github.com/dcgerard/updog/blob/master/R/multidog.R'><code>R/multidog.R</code></a></small>
    <div class="hidden name"><code>multidog.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is a convenience function that will run <code><a href='flexdog.html'>flexdog</a></code> over many SNPs.
Support is provided for parallel computing through the doParallel package.
This function has not been extensively tested. Please report any bugs to
<a href='http://github.com/dcgerard/updog/issues'>http://github.com/dcgerard/updog/issues</a>.</p>
    </div>

    <pre class="usage"><span class='fu'>multidog</span>(
  <span class='no'>refmat</span>,
  <span class='no'>sizemat</span>,
  <span class='no'>ploidy</span>,
  <span class='kw'>model</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"norm"</span>, <span class='st'>"hw"</span>, <span class='st'>"bb"</span>, <span class='st'>"s1"</span>, <span class='st'>"s1pp"</span>, <span class='st'>"f1"</span>, <span class='st'>"f1pp"</span>, <span class='st'>"flex"</span>, <span class='st'>"uniform"</span>,
    <span class='st'>"custom"</span>),
  <span class='kw'>nc</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>p1_id</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>p2_id</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bias_init</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>exp</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(-<span class='fl'>1</span>, -<span class='fl'>0.5</span>, <span class='fl'>0</span>, <span class='fl'>0.5</span>, <span class='fl'>1</span>)),
  <span class='kw'>prior_vec</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>refmat</th>
      <td><p>A matrix of reference read counts. The columns index
the individuals and the rows index the markers (SNPs). This matrix must have
rownames (for the names of the markers) and column names (for the names
of the individuals). These names must match the names in <code>sizemat</code>.</p></td>
    </tr>
    <tr>
      <th>sizemat</th>
      <td><p>A matrix of total read counts. The columns index
the individuals and the rows index the markers (SNPs). This matrix must have
rownames (for the names of the markers) and column names (for the names
of the individuals). These names must match the names in <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>The ploidy of the species. Assumed to be the same for each
individual.</p></td>
    </tr>
    <tr>
      <th>model</th>
      <td><p>What form should the prior (genotype distribution) take?
See Details for possible values.</p></td>
    </tr>
    <tr>
      <th>nc</th>
      <td><p>The number of computing cores to use. This should never be
more than the number of cores available in your computing environment.
You can determine the maximum number of available cores by running
<code><a href='https://rdrr.io/r/parallel/detectCores.html'>parallel::detectCores()</a></code> in R. But discuss how to run parallel
jobs with your admin if you are using <code>multidog()</code> on a
supercomputer.</p></td>
    </tr>
    <tr>
      <th>p1_id</th>
      <td><p>The ID of the first parent. This should be a character of
length 1. This should correspond to a single column name in <code>refmat</code>
and <code>sizemat</code>.</p></td>
    </tr>
    <tr>
      <th>p2_id</th>
      <td><p>The ID of the second parent. This should be a character of
length 1. This should correspond to a single column name in <code>refmat</code>
and <code>sizemat</code>.</p></td>
    </tr>
    <tr>
      <th>bias_init</th>
      <td><p>A vector of initial values for the bias parameter
over the multiple runs of <code><a href='flexdog_full.html'>flexdog_full</a>()</code>.</p></td>
    </tr>
    <tr>
      <th>prior_vec</th>
      <td><p>The pre-specified genotype distribution. Only used if
<code>model = "custom"</code> and must otherwise be <code>NULL</code>. If specified,
then it should be a vector of length <code>ploidy + 1</code> with
non-negative elements that sum to 1.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional parameters to pass to <code><a href='flexdog_full.html'>flexdog_full</a>()</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list-like object of two data frames.</p><dl'>
<dt><code>snpdf</code></dt><dd><p>A data frame containing properties of the SNPs (markers).
    The rows index the SNPs. The variables include:
    <dl'>
    <dt><code>snp</code></dt><dd><p>The name of the SNP (marker).</p></dd>
    <dt><code>bias</code></dt><dd><p>The estimated allele bias of the SNP.</p></dd>
    <dt><code>seq</code></dt><dd><p>The estimated sequencing error rate of the SNP.</p></dd>
    <dt><code>od</code></dt><dd><p>The estimated overdispersion parameter of the SNP.</p></dd>
    <dt><code>prop_mis</code></dt><dd><p>The estimated proportion of individuals
        misclassified in the SNP.</p></dd>
    <dt><code>num_iter</code></dt><dd><p>The number of iterations performed during
        the EM algorithm for that SNP.</p></dd>
    <dt><code>llike</code></dt><dd><p>The maximum marginal likelihood of the SNP.</p></dd>
    <dt><code>ploidy</code></dt><dd><p>The provided ploidy of the species.</p></dd>
    <dt><code>model</code></dt><dd><p>The provided model for the prior genotype
        distribution.</p></dd>
    <dt><code>p1ref</code></dt><dd><p>The user-provided reference read counts of parent 1.</p></dd>
    <dt><code>p1size</code></dt><dd><p>The user-provided total read counts of parent 1.</p></dd>
    <dt><code>p2ref</code></dt><dd><p>The user-provided reference read counts of parent 2.</p></dd>
    <dt><code>p2size</code></dt><dd><p>The user-provided total read counts of parent 2.</p></dd>
    <dt><code>Pr_k</code></dt><dd><p>The estimated frequency of individuals with genotype
        k, where k can be any integer between 0 and the ploidy level.</p></dd>
    <dt>Model specific parameter estimates</dt><dd><p>See the return value of
        <code>par</code> in the help page of <code><a href='flexdog.html'>flexdog</a></code>.</p></dd></dl></p></dd>
<dt><code>inddf</code></dt><dd><p>A data frame containing the properties of the
    individuals at each SNP. The variables include:
    <dl'>
    <dt><code>snp</code></dt><dd><p>The name of the SNP (marker).</p></dd>
    <dt><code>ind</code></dt><dd><p>The name of the individual.</p></dd>
    <dt><code>ref</code></dt><dd><p>The provided reference counts for that individual at
         that SNP.</p></dd>
    <dt><code>size</code></dt><dd><p>The provided total counts for that individual at
         that SNP.</p></dd>
    <dt><code>geno</code></dt><dd><p>The posterior mode genotype for that individual at
         that SNP. This is the estimated reference allele dosage for a
         given individual at a given SNP.</p></dd>
    <dt><code>postmean</code></dt><dd><p>The posterior mean genotype for that individual
         at that SNP. This is a continuous genotype estimate of the
         reference allele dosage for a given individual at a given SNP.</p></dd>
    <dt><code>maxpostprob</code></dt><dd><p>The maximum posterior probability. This
         is the posterior probability that the individual was genotyped
         correctly.</p></dd>
    <dt><code>Pr_k</code></dt><dd><p>The posterior probability that a given individual
         at a given SNP has genotype k, where k can vary from 0 to the
         ploidy level of the species.</p></dd>
    <dt><code>logL_k</code></dt><dd><p>The genotype <em>log</em>-likelihoods for dosage
         k for a given individual at a given SNP, where k can vary f
         rom 0 to the ploidy level of the species.</p></dd></dl></p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>You should format your reference counts and total read counts in two
separate matrices. The rows should index the markers (SNPs) and the
columns should index the individuals. Row names are how we ID the SNPs
and column names are how we ID the individuals, and so they are required
attributes.</p>
<p>If your data are in VCF files, I would recommend importing them using the
VariantAnnotation package from Bioconductor
<a href='https://bioconductor.org/packages/VariantAnnotation/'>https://bioconductor.org/packages/VariantAnnotation/</a>. It's a great
VCF parser.</p>
<p>See the details of <code><a href='flexdog.html'>flexdog</a></code> for the possible values of
<code>model</code>.</p>
<p>If <code>model = "f1"</code>, <code>model = "s1"</code>, <code>model = "f1pp"</code>
or <code>model = "s1pp"</code> then the user may
provide the individual ID for parent(s) via the <code>p1_id</code>
and <code>p2_id</code> arguments.</p>
<p>The output is a list containing two data frames. The first data frame,
called <code>snpdf</code>, contains information on each SNP, such as the allele bias
and the sequencing error rate. The second data frame, called <code>inddf</code>,
contains information on each individual at each SNP, such as the estimated
genotype and the posterior probability of being classified correctly.</p>
<p>Using an <code>nc</code> value greater than <code>1</code> will allow you to
run <code><a href='flexdog.html'>flexdog</a></code> in parallel. Only set <code>nc</code> greater than
<code>1</code> if you are sure you have access to the proper number of cores.
The upper bound on the value of <code>nc</code> you should try can be determined
by running <code><a href='https://rdrr.io/r/parallel/detectCores.html'>parallel::detectCores()</a></code> in R. Most admins of high
performance computing environments place limits on the number of cores
you can use at any one time. So if you are using <code>multidog()</code> on
a supercomputer, do not use <code><a href='https://rdrr.io/r/parallel/detectCores.html'>parallel::detectCores()</a></code> and discuss
with your admin how you can safely run parallel jobs.</p>
<p>SNPs that contain 0 reads (or all missing data) are entirely removed.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='flexdog.html'>flexdog</a>()</code>:For the underlying genotyping function.</p></li>
<li><p><code><a href='format_multidog.html'>format_multidog</a>()</code>:For converting the output
      of <code>multidog()</code> to a matrix.</p></li>
<li><p><code><a href='filter_snp.html'>filter_snp</a>()</code>:For filtering SNPs using the
      output of <code>multidog()</code>.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"uitdewilligen"</span>)
<span class='no'>mout</span> <span class='kw'>&lt;-</span> <span class='fu'>multidog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span>(<span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>),
                 <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span>(<span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>),
                 <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>ploidy</span>,
                 <span class='kw'>nc</span> <span class='kw'>=</span> <span class='fl'>2</span>)
<span class='no'>mout</span>$<span class='no'>inddf</span>
<span class='no'>mout</span>$<span class='no'>snpdf</span>
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


