<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Using correlation between individuals for genotyping. — mupdog • updog</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Using correlation between individuals for genotyping. — mupdog" />
<meta property="og:description" content="A method to genotype autopolyploids using GBS or RAD-seq like data by accounting
for correlations in the genotype distribution between the individuals.
This is a very experimental function. You probably want to use
multidog() instead." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/multidog.html">Genotyping Many SNP's with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Using correlation between individuals for genotyping.</h1>
    
    <div class="hidden name"><code>mupdog.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A method to genotype autopolyploids using GBS or RAD-seq like data by accounting
for correlations in the genotype distribution between the individuals.
This is a very experimental function. You probably want to use
<code><a href='multidog.html'>multidog</a>()</code> instead.</p>
    </div>

    <pre class="usage"><span class='fu'>mupdog</span>(
  <span class='no'>refmat</span>,
  <span class='no'>sizemat</span>,
  <span class='no'>ploidy</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>mean_bias</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>var_bias</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>mean_seq</span> <span class='kw'>=</span> -<span class='fl'>4.7</span>,
  <span class='kw'>var_seq</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>seq</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bias</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>od</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>allele_freq</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>inbreeding</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cor_mat</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>postmean</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>postvar</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>update_cor</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>update_inbreeding</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>update_allele_freq</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>num_core</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>update_method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Brent"</span>, <span class='st'>"L-BFGS-B"</span>),
  <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>()
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>refmat</th>
      <td><p>A matrix of reference counts.
The rows index the individuals and the columns index the SNPs.</p></td>
    </tr>
    <tr>
      <th>sizemat</th>
      <td><p>A matrix of total counts.
The rows index the individuals and the columns index the SNPs.
Should have the same dimensions as <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>The ploidy of the species. To estimate the ploidy,
re-run <code>mupdog</code> at various ploidy levels and choose the one with
the largest ELBO. This assumes that the ploidy is the same for
all individuals in the sample.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Should we print a lot of output <code>TRUE</code>
or not <code>FALSE</code>?</p></td>
    </tr>
    <tr>
      <th>mean_bias</th>
      <td><p>The prior mean of the log-bias. Defaults to 0
(no bias).</p></td>
    </tr>
    <tr>
      <th>var_bias</th>
      <td><p>The prior variance on the log-bias. Defaults to 1.
This roughly corresponds to likely bias values between 0.14 and
7.4. This is a far wider interval than what we observe in
practice, thus making this prior rather uninformative. We usually
observe bias values somewhere between 0.5 and 2.</p></td>
    </tr>
    <tr>
      <th>mean_seq</th>
      <td><p>The prior mean of the logit-sequencing-error-rate.
Defaults to -4.7. This corresponds to a sequencing error rate of
0.009.</p></td>
    </tr>
    <tr>
      <th>var_seq</th>
      <td><p>The prior variance of the logit-sequencing-error-rate.
Defaults to 1. This corresponds to likely values of 0.001 and
0.06. This upper bound is larger than what we would expect given
the current state of next-gen-sequencing technology.</p></td>
    </tr>
    <tr>
      <th>seq</th>
      <td><p>A vector of initial sequencing errors. Should be
the same length as the number of columns of <code>refmat</code>
(number of SNPs). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>bias</th>
      <td><p>A vector of initial bias parameters. Should be the
same length as the number of columns of <code>refmat</code>
(number of SNPs). Must be greater than 0.</p></td>
    </tr>
    <tr>
      <th>od</th>
      <td><p>A vector of initial overdispersion parameters.
Should be the same length as the number of columns of
<code>refmat</code> (number of SNPs). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>allele_freq</th>
      <td><p>A vector of initial allele frequencies.
Should be the same length as the number of columns of
<code>refmat</code> (number of SNPs). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>inbreeding</th>
      <td><p>A vector of initial individual-specific
inbreeding coefficients.
Should be the same length as the number of rows of
<code>refmat</code> (number of individuals). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>cor_mat</th>
      <td><p>Initial correlation matrix. Should have the same
number of columns/rows as the number of individuals.</p></td>
    </tr>
    <tr>
      <th>postmean</th>
      <td><p>Initial variational posterior means. Should
have the same dimensions as <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>postvar</th>
      <td><p>Initial posterior variances. Should have the
same dimensions as <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>update_cor</th>
      <td><p>A logical. Should we update the underlying
correlation matrix <code>TRUE</code> or not <code>FALSE</code>. It might be
unwise to set this to <code>TRUE</code> if you have more individuals
than SNPs.</p></td>
    </tr>
    <tr>
      <th>update_inbreeding</th>
      <td><p>A logical. Should we update the
inbreeding coefficients <code>TRUE</code>
or not <code>FALSE</code>?</p></td>
    </tr>
    <tr>
      <th>update_allele_freq</th>
      <td><p>A logical. Should we update the
allele frequencies <code>TRUE</code> or not <code>FALSE</code>?</p></td>
    </tr>
    <tr>
      <th>num_core</th>
      <td><p>The number of cores to use if you want to
run the optimization steps in parallel. If <code>num_core = 1</code>,
then the optimization step will not be run in parallel.</p></td>
    </tr>
    <tr>
      <th>update_method</th>
      <td><p>What generic optimizer should we use to update <code>allele_freq</code>
and <code>inbreeding</code>? Options are either <code>"Brent"</code> or <code>"L-BFGS-B"</code>.
See <code><a href='https://rdrr.io/r/stats/optim.html'>optim</a></code> for details on these optimizers.</p></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>A list of control parameters (<code>itermax</code>,
<code>obj_tol</code>).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with some or all of the following elements:</p><dl class='dl-horizontal'>
    <dt><code>map_dosage</code></dt><dd><p>A matrix of numerics containing
        the variational maximum-a-posterior (MAP) genotypes
        (allele dosages) for each individual at each SNP. Element
        (i, j) is the MAP genotype for individual i at SNP j.</p></dd>
    <dt><code>maxpostprob</code></dt><dd><p>A matrix of numerics containing
        the variational maximum posterior probabilities for each
        individual at each SNP. The (i, j)th element is the
        variational posterior
        probability that individual i is genotyped correctly at
        SNP j.</p></dd>
    <dt><code>postprob</code></dt><dd><p>A three-way array of numerics. Element (i, j, k) is
        the variational posterior probability that individual i has genotype
        k-1 at SNP j.</p></dd>
    <dt><code>seq</code></dt><dd><p>A vector of numerics. Element j is the estimated
        sequencing error rate for SNP j.</p></dd>
    <dt><code>bias</code></dt><dd><p>A vector of numerics. Element j is the estimated
        allelic bias for SNP j.</p></dd>
    <dt><code>od</code></dt><dd><p>A vector of numerics. Element j is the estimated
        overdispersion parameter for SNP j.</p></dd>
    <dt><code>allele_freq</code></dt><dd><p>A vector of numerics. Element j is the
        estimated allele-frequency for SNP j.</p></dd>
    <dt><code>inbreeding</code></dt><dd><p>A vector of numerics. Element i is the
        estimated inbreeding coefficient for individual i.</p></dd>
    <dt><code>cor_mat</code></dt><dd><p>A symmetric matrix of numerics. Element (i, j)
        is the estimated _latent_ correlation between individual
        i and individual j.</p></dd>
    <dt><code>postmean</code></dt><dd><p>A matrix of numerics. Element (i, j) is the
        variational posterior mean for individual i at SNP j.</p></dd>
    <dt><code>postvar</code></dt><dd><p>A matrix of numerics. Element (i, j) is the
        variational posterior variance for individual i at SNP j.</p></dd>
    <dt><code>input$refmat</code></dt><dd><p>A matrix of numerics.
        The inputted <code>refmat</code>.</p></dd>
    <dt><code>input$sizemat</code></dt><dd><p>A matrix of numerics. The inputted
        <code>sizemat</code>.</p></dd>
    <dt><code>input$ploidy</code></dt><dd><p>The inputted <code>ploidy</code>.</p></dd>
    <dt><code>obj</code></dt><dd><p>The maximized variational objective.</p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Blischak et al (2017) developed a genotyping approach for autopolyploids
that assumes a Balding-Nichols generative model (Balding and Nichols, 1997)
on the genotypes. Using a different generative model, Gerard et al (2018)
accounted for common issues in sequencing data ignored by previous researchers.
Mupdog unites and extends these two approaches:</p><ul>
<li><p>Unite: We account for locus-specific allele-bias, locus-specific sequencing error,
      and locus-specific overdispersion while marginally assuming a Balding-Nichols generative model on the genotypes.</p></li>
<li><p>Extend: We account for underlying correlations between the individuals using a Gaussian copula model.</p></li>
</ul><p>Mupdog uses a variational Bayes approach to estimate all parameters of interest and
the posterior probabilities of the genotypes for each individual at each locus.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
<ul>
<li><p>David J Balding and Richard A Nichols. <a href='http://dx.doi.org/10.1038/sj.hdy.6881750'>Significant genetic correlations among caucasians at forensic DNA loci</a>. <em>Heredity</em>, 78(6):583–589, 1997. doi: 10.1038/sj.hdy.6881750.</p></li>
<li><p>Paul D Blischak, Laura S Kubatko, and Andrea D Wolfe. <a href='http://dx.doi.org/10.1093/bioinformatics/btx587'>SNP genotyping and parameter estimation in polyploids using low-coverage sequencing data</a>. <em>Bioinformatics</em>, page btx587, 2017. doi: 10.1093/bioinformatics/btx587.</p></li>
<li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. doi: <a href='https://doi.org/10.1534/genetics.118.301468'>10.1534/genetics.118.301468</a>.</p></li>
</ul>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>uitdewilligen</span>)
<span class='no'>mout</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
               <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
               <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>ploidy</span>,
               <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
               <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))

<span class='co'>## Summaries of output</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span>(<span class='no'>mout</span>, <span class='fl'>4</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span>(<span class='no'>mout</span>$<span class='no'>bias</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span>(<span class='no'>mout</span>$<span class='no'>seq</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span>(<span class='no'>mout</span>$<span class='no'>od</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span>(<span class='no'>mout</span>$<span class='no'>inbreeding</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span>(<span class='no'>mout</span>$<span class='no'>allele_freq</span>)

<span class='co'>## mupdog can correctly estimate ploidy to be 4</span>
<span class='no'>mout2</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
                <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
                <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>,
                <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))

<span class='no'>mout6</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
                <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
                <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>6</span>,
                <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))

<span class='no'>mout8</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
                <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
                <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>8</span>,
                <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))

<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>mout2</span>$<span class='no'>obj</span>, <span class='no'>mout</span>$<span class='no'>obj</span>, <span class='no'>mout6</span>$<span class='no'>obj</span>, <span class='no'>mout8</span>$<span class='no'>obj</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(<span class='fl'>2</span>, <span class='fl'>8</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fl'>2</span>)
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"l"</span>, <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"ploidy"</span>, <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"objective"</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>David Gerard</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


