<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Multi-SNP updog. — mupdog • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Multi-SNP updog. — mupdog" />

<meta property="og:description" content="A method to genotype autopolyploids using GBS or RAD-seq like data by accounting
for correlations in the genotype distribution between the individuals." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Multi-SNP updog.</h1>
    
    <div class="hidden name"><code>mupdog.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>A method to genotype autopolyploids using GBS or RAD-seq like data by accounting
for correlations in the genotype distribution between the individuals.</p>
    
    </div>

    <pre class="usage"><span class='fu'>mupdog</span>(<span class='no'>refmat</span>, <span class='no'>sizemat</span>, <span class='no'>ploidy</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>mean_bias</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>var_bias</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>mean_seq</span> <span class='kw'>=</span> -<span class='fl'>4.7</span>, <span class='kw'>var_seq</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>seq</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bias</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>od</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>allele_freq</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>inbreeding</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cor_mat</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>postmean</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>postvar</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>update_cor</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_inbreeding</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>update_allele_freq</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>num_core</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>update_method</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"Brent"</span>,
  <span class='st'>"L-BFGS-B"</span>), <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'>list</span>())</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>refmat</th>
      <td><p>A matrix of reference counts.
The rows index the individuals and the columns index the SNPs.</p></td>
    </tr>
    <tr>
      <th>sizemat</th>
      <td><p>A matrix of total counts.
The rows index the individuals and the columns index the SNPs.
Should have the same dimensions as <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>The ploidy of the species. To estimate the ploidy,
re-run <code>mupdog</code> at various ploidy levels and choose the one with
the largest ELBO. This assumes that the ploidy is the same for
all individuals in the sample.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Should we print a lot of output <code>TRUE</code>
or not <code>FALSE</code>?</p></td>
    </tr>
    <tr>
      <th>mean_bias</th>
      <td><p>The prior mean of the log-bias. Defaults to 0
(no bias).</p></td>
    </tr>
    <tr>
      <th>var_bias</th>
      <td><p>The prior variance on the log-bias. Defaults to 1.
This roughly corresponds to likely bias values between 0.14 and
7.4. This is a far wider interval than what we observe in
practice, thus making this prior rather uninformative. We usually
observe bias values somewhere between 0.5 and 2.</p></td>
    </tr>
    <tr>
      <th>mean_seq</th>
      <td><p>The prior mean of the logit-sequencing-error-rate.
Defaults to -4.7. This corresponds to a sequencing error rate of
0.009.</p></td>
    </tr>
    <tr>
      <th>var_seq</th>
      <td><p>The prior variance of the logit-sequencing-error-rate.
Defaults to 1. This corresponds to likely values of 0.001 and
0.06. This upper bound is larger than what we would expect given
the current state of next-gen-sequencing technology.</p></td>
    </tr>
    <tr>
      <th>seq</th>
      <td><p>A vector of initial sequencing errors. Should be
the same length as the number of columns of <code>refmat</code>
(number of SNPs). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>bias</th>
      <td><p>A vector of initial bias parameters. Should be the
same length as the number of columns of <code>refmat</code>
(number of SNPs). Must be greater than 0.</p></td>
    </tr>
    <tr>
      <th>od</th>
      <td><p>A vector of initial overdispersion parameters.
Should be the same length as the number of columns of
<code>refmat</code> (number of SNPs). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>allele_freq</th>
      <td><p>A vector of initial allele frequencies.
Should be the same length as the number of columns of
<code>refmat</code> (number of SNPs). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>inbreeding</th>
      <td><p>A vector of initial individual-specific
inbreeding coefficients.
Should be the same length as the number of rows of
<code>refmat</code> (number of individuals). Must be between 0 and 1.</p></td>
    </tr>
    <tr>
      <th>cor_mat</th>
      <td><p>Initial correlation matrix. Should have the same
number of columns/rows as the number of individuals.</p></td>
    </tr>
    <tr>
      <th>postmean</th>
      <td><p>Initial variational posterior means. Should
have the same dimensions as <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>postvar</th>
      <td><p>Initial posterior variances. Should have the
same dimensions as <code>refmat</code>.</p></td>
    </tr>
    <tr>
      <th>update_cor</th>
      <td><p>A logical. Should we update the underlying
correlation matrix <code>TRUE</code> or not <code>FALSE</code>. It might be
unwise to set this to <code>TRUE</code> if you have more individuals
than SNPs.</p></td>
    </tr>
    <tr>
      <th>update_inbreeding</th>
      <td><p>A logical. Should we update the
inbreeding coefficients <code>TRUE</code>
or not <code>FALSE</code>?</p></td>
    </tr>
    <tr>
      <th>update_allele_freq</th>
      <td><p>A logical. Should we update the
allele frequencies <code>TRUE</code> or not <code>FALSE</code>?</p></td>
    </tr>
    <tr>
      <th>num_core</th>
      <td><p>The number of cores to use if you want to
run the optimization steps in parallel. If <code>num_core = 1</code>,
then the optimization step will not be run in parallel.</p></td>
    </tr>
    <tr>
      <th>update_method</th>
      <td><p>What generic optimizer should we use to update <code>allele_freq</code>
and <code>inbreeding</code>? Options are either <code>"Brent"</code> or <code>"L-BFGS-B"</code>.
See <code><a href='http://www.rdocumentation.org/packages/stats/topics/optim'>optim</a></code> for details on these optimizers.</p></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>A list of control parameters (<code>itermax</code>,
<code>obj_tol</code>).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with some or all of the following elements:</p><dl class='dl-horizontal'>
    <dt><code>map_dosage</code></dt><dd><p>A matrix of numerics containing
        the variational maximum-a-posterior (MAP) genotypes
        (allele dosages) for each individual at each SNP. Element
        (i, j) is the MAP genotype for individual i at SNP j.</p></dd>
    <dt><code>maxpostprob</code></dt><dd><p>A matrix of numerics containing
        the variational maximum posterior probabilities for each
        individual at each SNP. The (i, j)th element is the
        variational posterior
        probability that individual i is genotyped correctly at
        SNP j.</p></dd>
    <dt><code>postprob</code></dt><dd><p>A three-way array of numerics. Element (i, j, k) is
        the variational posterior probability that individual i has genotype
        k-1 at SNP j.</p></dd>
    <dt><code>seq</code></dt><dd><p>A vector of numerics. Element j is the estimated
        sequencing error rate for SNP j.</p></dd>
    <dt><code>bias</code></dt><dd><p>A vector of numerics. Element j is the estimated
        allelic bias for SNP j.</p></dd>
    <dt><code>od</code></dt><dd><p>A vector of numerics. Element j is the estimated
        overdispersion parameter for SNP j.</p></dd>
    <dt><code>allele_freq</code></dt><dd><p>A vector of numerics. Element j is the
        estimated allele-frequency for SNP j.</p></dd>
    <dt><code>inbreeding</code></dt><dd><p>A vector of numerics. Element i is the
        estimated inbreeding coefficient for individual i.</p></dd>
    <dt><code>cor_mat</code></dt><dd><p>A symmetric matrix of numerics. Element (i, j)
        is the estimated _latent_ correlation between individual
        i and individual j.</p></dd>
    <dt><code>postmean</code></dt><dd><p>A matrix of numerics. Element (i, j) is the
        variational posterior mean for individual i at SNP j.</p></dd>
    <dt><code>postvar</code></dt><dd><p>A matrix of numerics. Element (i, j) is the
        variational posterior variance for individual i at SNP j.</p></dd>
    <dt><code>input$refmat</code></dt><dd><p>A matrix of numerics.
        The inputted <code>refmat</code>.</p></dd>
    <dt><code>input$sizemat</code></dt><dd><p>A matrix of numerics. The inputted
        <code>sizemat</code>.</p></dd>
    <dt><code>input$ploidy</code></dt><dd><p>The inputted <code>ploidy</code>.</p></dd>
    <dt><code>obj</code></dt><dd><p>The maximized variational objective.</p></dd>
</dl>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Blischak et al (2017) developed a genotyping approach for autopolyploids
that assumes a Balding-Nichols generative model (Balding and Nichols, 1997)
on the genotypes. Using a different generative model, Gerard et al (2018)
accounted for common issues in sequencing data ignored by previous researchers.
Mupdog unites and extends these two approaches:</p><ul>
<li><p>Unite: We account for locus-specific allele-bias, locus-specific sequencing error,
      and locus-specific overdispersion while marginally assuming a Balding-Nichols generative model on the genotypes.</p></li>
<li><p>Extend: We account for underlying correlations between the individuals using a Gaussian copula model.</p></li>
</ul><p>Mupdog uses a variational Bayes approach to estimate all parameters of interest and
the posterior probabilities of the genotypes for each individual at each locus.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <ul>
<li><p>David J Balding and Richard A Nichols. <a href='http://dx.doi.org/10.1038/sj.hdy.6881750'>Significant genetic correlations among caucasians at forensic DNA loci</a>. Heredity, 78(6):583–589, 1997. doi: 10.1038/sj.hdy.6881750.</p></li>
<li><p>Paul D Blischak, Laura S Kubatko, and Andrea D Wolfe. <a href='http://dx.doi.org/10.1093/bioinformatics/btx587'>SNP genotyping and parameter estimation in polyploids using low-coverage sequencing data</a>. Bioinformatics, page btx587, 2017. doi: 10.1093/bioinformatics/btx587.</p></li>
<li><p>Gerard, D., Ferrao, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. *Genetics*, 210(3), 789-807. doi: [10.1534/genetics.118.301468](https://doi.org/10.1534/genetics.118.301468).</p></li>
</ul>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>uitdewilligen</span>)
<span class='no'>mout</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
               <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
               <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>ploidy</span>,
               <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
               <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))</div><div class='output co'>#&gt;  iteration: 1 
#&gt;  objective: -4079.939 
#&gt;        err: Inf 
#&gt; 
#&gt;  iteration: 2 
#&gt;  objective: -3713.483 
#&gt;        err: 0.09868254 
#&gt; 
#&gt;  iteration: 3 
#&gt;  objective: -3614.958 
#&gt;        err: 0.02725493 
#&gt; 
#&gt;  iteration: 4 
#&gt;  objective: -3559.182 
#&gt;        err: 0.01567103 
#&gt; 
#&gt;  iteration: 5 
#&gt;  objective: -3522.352 
#&gt;        err: 0.01045596 
#&gt; 
#&gt;  iteration: 6 
#&gt;  objective: -3499.16 
#&gt;        err: 0.00662798 
#&gt; 
#&gt;  iteration: 7 
#&gt;  objective: -3485.76 
#&gt;        err: 0.003844147 
#&gt; 
#&gt;  iteration: 8 
#&gt;  objective: -3477.963 
#&gt;        err: 0.002242003 
#&gt; 
#&gt;  iteration: 9 
#&gt;  objective: -3472.599 
#&gt;        err: 0.001544479 
#&gt; 
#&gt;  iteration: 10 
#&gt;  objective: -3468.656 
#&gt;        err: 0.001136844 
#&gt; 
#&gt;  iteration: 11 
#&gt;  objective: -3465.761 
#&gt;        err: 0.0008353654 
#&gt; 
#&gt;  iteration: 12 
#&gt;  objective: -3463.696 
#&gt;        err: 0.0005961956 
#&gt; 
#&gt;  iteration: 13 
#&gt;  objective: -3461.904 
#&gt;        err: 0.0005176985 
#&gt; 
#&gt;  iteration: 14 
#&gt;  objective: -3460.361 
#&gt;        err: 0.0004456278 
#&gt; 
#&gt;  iteration: 15 
#&gt;  objective: -3459.117 
#&gt;        err: 0.0003597917 
#&gt; 
#&gt;  iteration: 16 
#&gt;  objective: -3458.061 
#&gt;        err: 0.0003054564 
#&gt; 
#&gt;  iteration: 17 
#&gt;  objective: -3457.082 
#&gt;        err: 0.0002831857 
#&gt; 
#&gt;  iteration: 18 
#&gt;  objective: -3456.205 
#&gt;        err: 0.000253511 
#&gt; 
#&gt;  iteration: 19 
#&gt;  objective: -3455.446 
#&gt;        err: 0.0002198243 
#&gt; 
#&gt;  iteration: 20 
#&gt;  objective: -3454.795 
#&gt;        err: 0.0001883458 
#&gt; 
#&gt;  iteration: 21 
#&gt;  objective: -3454.239 
#&gt;        err: 0.0001611014 
#&gt; 
#&gt;  iteration: 22 
#&gt;  objective: -3453.724 
#&gt;        err: 0.0001489191 
#&gt; 
#&gt;  iteration: 23 
#&gt;  objective: -3453.251 
#&gt;        err: 0.0001371688 
#&gt; 
#&gt;  iteration: 24 
#&gt;  objective: -3452.806 
#&gt;        err: 0.000128903 
#&gt; 
#&gt;  iteration: 25 
#&gt;  objective: -3452.369 
#&gt;        err: 0.0001264393 
#&gt; 
#&gt;  iteration: 26 
#&gt;  objective: -3451.951 
#&gt;        err: 0.0001210796 
#&gt; 
#&gt;  iteration: 27 
#&gt;  objective: -3451.542 
#&gt;        err: 0.0001186601 
#&gt; 
#&gt;  iteration: 28 
#&gt;  objective: -3451.147 
#&gt;        err: 0.000114472 
#&gt; 
#&gt;  iteration: 29 
#&gt;  objective: -3450.754 
#&gt;        err: 0.0001137623 
#&gt; 
#&gt;  iteration: 30 
#&gt;  objective: -3450.377 
#&gt;        err: 0.0001092022 
#&gt; 
#&gt;  iteration: 31 
#&gt;  objective: -3450.001 
#&gt;        err: 0.0001090136 
#&gt; 
#&gt;  iteration: 32 
#&gt;  objective: -3449.637 
#&gt;        err: 0.0001054427 
#&gt; 
#&gt;  iteration: 33 
#&gt;  objective: -3449.278 
#&gt;        err: 0.0001042686 
#&gt; 
#&gt;  iteration: 34 
#&gt;  objective: -3448.928 
#&gt;        err: 0.0001014436 
#&gt; 
#&gt;  iteration: 35 
#&gt;  objective: -3448.583 
#&gt;        err: 9.991288e-05 
#&gt; </div><div class='input'>
<span class='co'>## Summaries of output</span>
<span class='fu'>plot</span>(<span class='no'>mout</span>, <span class='fl'>4</span>)</div><div class='img'><img src='mupdog-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>hist</span>(<span class='no'>mout</span>$<span class='no'>bias</span>)</div><div class='img'><img src='mupdog-2.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>hist</span>(<span class='no'>mout</span>$<span class='no'>seq</span>)</div><div class='img'><img src='mupdog-3.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>hist</span>(<span class='no'>mout</span>$<span class='no'>od</span>)</div><div class='img'><img src='mupdog-4.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>hist</span>(<span class='no'>mout</span>$<span class='no'>inbreeding</span>)</div><div class='img'><img src='mupdog-5.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>hist</span>(<span class='no'>mout</span>$<span class='no'>allele_freq</span>)</div><div class='img'><img src='mupdog-6.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## mupdog can correctly estimate ploidy to be 4</span>
<span class='no'>mout2</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
                <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
                <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>,
                <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))</div><div class='output co'>#&gt;  iteration: 1 
#&gt;  objective: -4975.229 
#&gt;        err: Inf 
#&gt; 
#&gt;  iteration: 2 
#&gt;  objective: -4407.658 
#&gt;        err: 0.1287691 
#&gt; 
#&gt;  iteration: 3 
#&gt;  objective: -4221.376 
#&gt;        err: 0.04412832 
#&gt; 
#&gt;  iteration: 4 
#&gt;  objective: -4091.237 
#&gt;        err: 0.03180931 
#&gt; 
#&gt;  iteration: 5 
#&gt;  objective: -4002.692 
#&gt;        err: 0.02212127 
#&gt; 
#&gt;  iteration: 6 
#&gt;  objective: -3941.859 
#&gt;        err: 0.01543257 
#&gt; 
#&gt;  iteration: 7 
#&gt;  objective: -3899.622 
#&gt;        err: 0.010831 
#&gt; 
#&gt;  iteration: 8 
#&gt;  objective: -3865.909 
#&gt;        err: 0.008720518 
#&gt; 
#&gt;  iteration: 9 
#&gt;  objective: -3841.693 
#&gt;        err: 0.006303706 
#&gt; 
#&gt;  iteration: 10 
#&gt;  objective: -3823.751 
#&gt;        err: 0.004692072 
#&gt; 
#&gt;  iteration: 11 
#&gt;  objective: -3808.765 
#&gt;        err: 0.003934673 
#&gt; 
#&gt;  iteration: 12 
#&gt;  objective: -3795.382 
#&gt;        err: 0.003526192 
#&gt; 
#&gt;  iteration: 13 
#&gt;  objective: -3783.53 
#&gt;        err: 0.003132422 
#&gt; 
#&gt;  iteration: 14 
#&gt;  objective: -3773.925 
#&gt;        err: 0.00254516 
#&gt; 
#&gt;  iteration: 15 
#&gt;  objective: -3764.818 
#&gt;        err: 0.002418945 
#&gt; 
#&gt;  iteration: 16 
#&gt;  objective: -3756.938 
#&gt;        err: 0.0020974 
#&gt; 
#&gt;  iteration: 17 
#&gt;  objective: -3749.826 
#&gt;        err: 0.001896631 
#&gt; 
#&gt;  iteration: 18 
#&gt;  objective: -3745.089 
#&gt;        err: 0.001264927 
#&gt; 
#&gt;  iteration: 19 
#&gt;  objective: -3740.855 
#&gt;        err: 0.001131873 
#&gt; 
#&gt;  iteration: 20 
#&gt;  objective: -3737.208 
#&gt;        err: 0.0009757388 
#&gt; 
#&gt;  iteration: 21 
#&gt;  objective: -3733.807 
#&gt;        err: 0.000910844 
#&gt; 
#&gt;  iteration: 22 
#&gt;  objective: -3730.064 
#&gt;        err: 0.001003434 
#&gt; 
#&gt;  iteration: 23 
#&gt;  objective: -3723.976 
#&gt;        err: 0.001634913 
#&gt; 
#&gt;  iteration: 24 
#&gt;  objective: -3718.855 
#&gt;        err: 0.001376926 
#&gt; 
#&gt;  iteration: 25 
#&gt;  objective: -3716.162 
#&gt;        err: 0.0007247415 
#&gt; 
#&gt;  iteration: 26 
#&gt;  objective: -3713.715 
#&gt;        err: 0.0006590882 
#&gt; 
#&gt;  iteration: 27 
#&gt;  objective: -3711.555 
#&gt;        err: 0.0005819355 
#&gt; 
#&gt;  iteration: 28 
#&gt;  objective: -3709.702 
#&gt;        err: 0.00049932 
#&gt; 
#&gt;  iteration: 29 
#&gt;  objective: -3707.908 
#&gt;        err: 0.0004839697 
#&gt; 
#&gt;  iteration: 30 
#&gt;  objective: -3706.376 
#&gt;        err: 0.000413252 
#&gt; 
#&gt;  iteration: 31 
#&gt;  objective: -3704.964 
#&gt;        err: 0.0003812164 
#&gt; 
#&gt;  iteration: 32 
#&gt;  objective: -3703.481 
#&gt;        err: 0.000400346 
#&gt; 
#&gt;  iteration: 33 
#&gt;  objective: -3701.749 
#&gt;        err: 0.000467884 
#&gt; 
#&gt;  iteration: 34 
#&gt;  objective: -3700.095 
#&gt;        err: 0.0004470067 
#&gt; 
#&gt;  iteration: 35 
#&gt;  objective: -3697.83 
#&gt;        err: 0.0006125821 
#&gt; 
#&gt;  iteration: 36 
#&gt;  objective: -3693.951 
#&gt;        err: 0.001050069 
#&gt; 
#&gt;  iteration: 37 
#&gt;  objective: -3690.656 
#&gt;        err: 0.0008929257 
#&gt; 
#&gt;  iteration: 38 
#&gt;  objective: -3688.832 
#&gt;        err: 0.0004942834 
#&gt; 
#&gt;  iteration: 39 
#&gt;  objective: -3686.954 
#&gt;        err: 0.0005094158 
#&gt; 
#&gt;  iteration: 40 
#&gt;  objective: -3685.617 
#&gt;        err: 0.0003626849 
#&gt; 
#&gt;  iteration: 41 
#&gt;  objective: -3682.717 
#&gt;        err: 0.0007875322 
#&gt; 
#&gt;  iteration: 42 
#&gt;  objective: -3680.622 
#&gt;        err: 0.0005691395 
#&gt; 
#&gt;  iteration: 43 
#&gt;  objective: -3679.288 
#&gt;        err: 0.0003626018 
#&gt; 
#&gt;  iteration: 44 
#&gt;  objective: -3677.856 
#&gt;        err: 0.0003894257 
#&gt; 
#&gt;  iteration: 45 
#&gt;  objective: -3676.74 
#&gt;        err: 0.0003034048 
#&gt; 
#&gt;  iteration: 46 
#&gt;  objective: -3675.622 
#&gt;        err: 0.0003043207 
#&gt; 
#&gt;  iteration: 47 
#&gt;  objective: -3674.278 
#&gt;        err: 0.0003657357 
#&gt; 
#&gt;  iteration: 48 
#&gt;  objective: -3673.465 
#&gt;        err: 0.0002213166 
#&gt; 
#&gt;  iteration: 49 
#&gt;  objective: -3672.644 
#&gt;        err: 0.0002234397 
#&gt; 
#&gt;  iteration: 50 
#&gt;  objective: -3671.943 
#&gt;        err: 0.0001909096 
#&gt; 
#&gt;  iteration: 51 
#&gt;  objective: -3671.324 
#&gt;        err: 0.000168613 
#&gt; 
#&gt;  iteration: 52 
#&gt;  objective: -3670.725 
#&gt;        err: 0.0001632885 
#&gt; 
#&gt;  iteration: 53 
#&gt;  objective: -3669.596 
#&gt;        err: 0.0003075495 
#&gt; 
#&gt;  iteration: 54 
#&gt;  objective: -3667.427 
#&gt;        err: 0.0005915037 
#&gt; 
#&gt;  iteration: 55 
#&gt;  objective: -3666.232 
#&gt;        err: 0.000325877 
#&gt; 
#&gt;  iteration: 56 
#&gt;  objective: -3665.615 
#&gt;        err: 0.0001683052 
#&gt; 
#&gt;  iteration: 57 
#&gt;  objective: -3665.07 
#&gt;        err: 0.0001487262 
#&gt; 
#&gt;  iteration: 58 
#&gt;  objective: -3664.528 
#&gt;        err: 0.0001478492 
#&gt; 
#&gt;  iteration: 59 
#&gt;  objective: -3664.042 
#&gt;        err: 0.000132772 
#&gt; 
#&gt;  iteration: 60 
#&gt;  objective: -3663.573 
#&gt;        err: 0.000128081 
#&gt; 
#&gt;  iteration: 61 
#&gt;  objective: -3663.116 
#&gt;        err: 0.0001247343 
#&gt; 
#&gt;  iteration: 62 
#&gt;  objective: -3662.694 
#&gt;        err: 0.0001150323 
#&gt; 
#&gt;  iteration: 63 
#&gt;  objective: -3662.323 
#&gt;        err: 0.0001015479 
#&gt; 
#&gt;  iteration: 64 
#&gt;  objective: -3661.926 
#&gt;        err: 0.0001083876 
#&gt; 
#&gt;  iteration: 65 
#&gt;  objective: -3661.578 
#&gt;        err: 9.505013e-05 
#&gt; </div><div class='input'>
<span class='no'>mout6</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
                <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
                <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>6</span>,
                <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))</div><div class='output co'>#&gt;  iteration: 1 
#&gt;  objective: -3929.194 
#&gt;        err: Inf 
#&gt; 
#&gt;  iteration: 2 
#&gt;  objective: -3727.8 
#&gt;        err: 0.05402482 
#&gt; 
#&gt;  iteration: 3 
#&gt;  objective: -3678.794 
#&gt;        err: 0.01332121 
#&gt; 
#&gt;  iteration: 4 
#&gt;  objective: -3653.876 
#&gt;        err: 0.006819692 
#&gt; 
#&gt;  iteration: 5 
#&gt;  objective: -3639.264 
#&gt;        err: 0.004015172 
#&gt; 
#&gt;  iteration: 6 
#&gt;  objective: -3628.725 
#&gt;        err: 0.002904176 
#&gt; 
#&gt;  iteration: 7 
#&gt;  objective: -3620.386 
#&gt;        err: 0.00230338 
#&gt; 
#&gt;  iteration: 8 
#&gt;  objective: -3613.08 
#&gt;        err: 0.002022231 
#&gt; 
#&gt;  iteration: 9 
#&gt;  objective: -3606.442 
#&gt;        err: 0.001840657 
#&gt; 
#&gt;  iteration: 10 
#&gt;  objective: -3600.535 
#&gt;        err: 0.001640433 
#&gt; 
#&gt;  iteration: 11 
#&gt;  objective: -3595.212 
#&gt;        err: 0.001480693 
#&gt; 
#&gt;  iteration: 12 
#&gt;  objective: -3590.791 
#&gt;        err: 0.001231184 
#&gt; 
#&gt;  iteration: 13 
#&gt;  objective: -3587.49 
#&gt;        err: 0.0009201664 
#&gt; 
#&gt;  iteration: 14 
#&gt;  objective: -3585.097 
#&gt;        err: 0.0006672755 
#&gt; 
#&gt;  iteration: 15 
#&gt;  objective: -3583.081 
#&gt;        err: 0.0005626644 
#&gt; 
#&gt;  iteration: 16 
#&gt;  objective: -3581.219 
#&gt;        err: 0.0005199719 
#&gt; 
#&gt;  iteration: 17 
#&gt;  objective: -3579.371 
#&gt;        err: 0.0005162283 
#&gt; 
#&gt;  iteration: 18 
#&gt;  objective: -3576.957 
#&gt;        err: 0.0006749337 
#&gt; 
#&gt;  iteration: 19 
#&gt;  objective: -3574.238 
#&gt;        err: 0.0007608579 
#&gt; 
#&gt;  iteration: 20 
#&gt;  objective: -3572.143 
#&gt;        err: 0.0005865616 
#&gt; 
#&gt;  iteration: 21 
#&gt;  objective: -3570.313 
#&gt;        err: 0.0005125002 
#&gt; 
#&gt;  iteration: 22 
#&gt;  objective: -3568.387 
#&gt;        err: 0.0005397542 
#&gt; 
#&gt;  iteration: 23 
#&gt;  objective: -3566.608 
#&gt;        err: 0.0004986471 
#&gt; 
#&gt;  iteration: 24 
#&gt;  objective: -3565.357 
#&gt;        err: 0.0003507926 
#&gt; 
#&gt;  iteration: 25 
#&gt;  objective: -3564.272 
#&gt;        err: 0.0003045195 
#&gt; 
#&gt;  iteration: 26 
#&gt;  objective: -3563.101 
#&gt;        err: 0.0003287296 
#&gt; 
#&gt;  iteration: 27 
#&gt;  objective: -3561.886 
#&gt;        err: 0.0003410293 
#&gt; 
#&gt;  iteration: 28 
#&gt;  objective: -3560.713 
#&gt;        err: 0.0003294316 
#&gt; 
#&gt;  iteration: 29 
#&gt;  objective: -3559.514 
#&gt;        err: 0.0003369128 
#&gt; 
#&gt;  iteration: 30 
#&gt;  objective: -3558.189 
#&gt;        err: 0.0003722555 
#&gt; 
#&gt;  iteration: 31 
#&gt;  objective: -3556.818 
#&gt;        err: 0.0003854869 
#&gt; 
#&gt;  iteration: 32 
#&gt;  objective: -3555.945 
#&gt;        err: 0.0002456401 
#&gt; 
#&gt;  iteration: 33 
#&gt;  objective: -3555.376 
#&gt;        err: 0.0001599917 
#&gt; 
#&gt;  iteration: 34 
#&gt;  objective: -3554.881 
#&gt;        err: 0.0001390875 
#&gt; 
#&gt;  iteration: 35 
#&gt;  objective: -3554.526 
#&gt;        err: 9.987911e-05 
#&gt; </div><div class='input'>
<span class='no'>mout8</span> <span class='kw'>&lt;-</span> <span class='fu'>mupdog</span>(<span class='kw'>refmat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>refmat</span>,
                <span class='kw'>sizemat</span> <span class='kw'>=</span> <span class='no'>uitdewilligen</span>$<span class='no'>sizemat</span>,
                <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>8</span>,
                <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                <span class='kw'>control</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>obj_tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>))</div><div class='output co'>#&gt;  iteration: 1 
#&gt;  objective: -3871.083 
#&gt;        err: Inf 
#&gt; 
#&gt;  iteration: 2 
#&gt;  objective: -3729.474 
#&gt;        err: 0.03797014 
#&gt; 
#&gt;  iteration: 3 
#&gt;  objective: -3693.045 
#&gt;        err: 0.009864317 
#&gt; 
#&gt;  iteration: 4 
#&gt;  objective: -3671.325 
#&gt;        err: 0.005916019 
#&gt; 
#&gt;  iteration: 5 
#&gt;  objective: -3657.241 
#&gt;        err: 0.003851021 
#&gt; 
#&gt;  iteration: 6 
#&gt;  objective: -3647.551 
#&gt;        err: 0.002656546 
#&gt; 
#&gt;  iteration: 7 
#&gt;  objective: -3640.437 
#&gt;        err: 0.001954115 
#&gt; 
#&gt;  iteration: 8 
#&gt;  objective: -3634.97 
#&gt;        err: 0.001504072 
#&gt; 
#&gt;  iteration: 9 
#&gt;  objective: -3630.248 
#&gt;        err: 0.001300586 
#&gt; 
#&gt;  iteration: 10 
#&gt;  objective: -3626.167 
#&gt;        err: 0.001125631 
#&gt; 
#&gt;  iteration: 11 
#&gt;  objective: -3622.552 
#&gt;        err: 0.000997736 
#&gt; 
#&gt;  iteration: 12 
#&gt;  objective: -3619.348 
#&gt;        err: 0.0008854201 
#&gt; 
#&gt;  iteration: 13 
#&gt;  objective: -3616.468 
#&gt;        err: 0.0007962961 
#&gt; 
#&gt;  iteration: 14 
#&gt;  objective: -3614.055 
#&gt;        err: 0.0006675477 
#&gt; 
#&gt;  iteration: 15 
#&gt;  objective: -3612.026 
#&gt;        err: 0.0005619301 
#&gt; 
#&gt;  iteration: 16 
#&gt;  objective: -3610.226 
#&gt;        err: 0.0004985697 
#&gt; 
#&gt;  iteration: 17 
#&gt;  objective: -3608.607 
#&gt;        err: 0.0004485358 
#&gt; 
#&gt;  iteration: 18 
#&gt;  objective: -3607.354 
#&gt;        err: 0.0003473327 
#&gt; 
#&gt;  iteration: 19 
#&gt;  objective: -3606.309 
#&gt;        err: 0.0002899184 
#&gt; 
#&gt;  iteration: 20 
#&gt;  objective: -3605.38 
#&gt;        err: 0.0002576682 
#&gt; 
#&gt;  iteration: 21 
#&gt;  objective: -3604.538 
#&gt;        err: 0.0002335943 
#&gt; 
#&gt;  iteration: 22 
#&gt;  objective: -3603.734 
#&gt;        err: 0.000222971 
#&gt; 
#&gt;  iteration: 23 
#&gt;  objective: -3602.977 
#&gt;        err: 0.0002102571 
#&gt; 
#&gt;  iteration: 24 
#&gt;  objective: -3602.248 
#&gt;        err: 0.0002023532 
#&gt; 
#&gt;  iteration: 25 
#&gt;  objective: -3601.508 
#&gt;        err: 0.0002054489 
#&gt; 
#&gt;  iteration: 26 
#&gt;  objective: -3600.764 
#&gt;        err: 0.0002065221 
#&gt; 
#&gt;  iteration: 27 
#&gt;  objective: -3600.077 
#&gt;        err: 0.0001908426 
#&gt; 
#&gt;  iteration: 28 
#&gt;  objective: -3599.304 
#&gt;        err: 0.0002148535 
#&gt; 
#&gt;  iteration: 29 
#&gt;  objective: -3598.461 
#&gt;        err: 0.0002340623 
#&gt; 
#&gt;  iteration: 30 
#&gt;  objective: -3597.529 
#&gt;        err: 0.0002592149 
#&gt; 
#&gt;  iteration: 31 
#&gt;  objective: -3596.771 
#&gt;        err: 0.0002107296 
#&gt; 
#&gt;  iteration: 32 
#&gt;  objective: -3596.119 
#&gt;        err: 0.0001811932 
#&gt; 
#&gt;  iteration: 33 
#&gt;  objective: -3595.59 
#&gt;        err: 0.0001472328 
#&gt; 
#&gt;  iteration: 34 
#&gt;  objective: -3595.153 
#&gt;        err: 0.0001215347 
#&gt; 
#&gt;  iteration: 35 
#&gt;  objective: -3594.76 
#&gt;        err: 0.0001094471 
#&gt; 
#&gt;  iteration: 36 
#&gt;  objective: -3594.41 
#&gt;        err: 9.724829e-05 
#&gt; </div><div class='input'>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='no'>mout2</span>$<span class='no'>obj</span>, <span class='no'>mout</span>$<span class='no'>obj</span>, <span class='no'>mout6</span>$<span class='no'>obj</span>, <span class='no'>mout8</span>$<span class='no'>obj</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>seq</span>(<span class='fl'>2</span>, <span class='fl'>8</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fl'>2</span>)
<span class='fu'>plot</span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"l"</span>, <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"ploidy"</span>, <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"objective"</span>)</div><div class='img'><img src='mupdog-7.png' alt='' width='700' height='433' /></div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

