<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate oracle misclassification error rate. — oracle_mis • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Calculate oracle misclassification error rate. — oracle_mis" />

<meta property="og:description" content="Given perfect knowledge of the data generating parameters,
oracle_mis calculates the misclassification error
rate, where the error rate is taken over both the data generation
process and the allele-distribution.
This is an ideal level of the misclassification error rate and
any real method will have a larger rate than this. This is a useful
approximation when you have a lot of individuals." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate oracle misclassification error rate.</h1>
    
    <div class="hidden name"><code>oracle_mis.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Given perfect knowledge of the data generating parameters,
<code>oracle_mis</code> calculates the misclassification error
rate, where the error rate is taken over both the data generation
process and the allele-distribution.
This is an ideal level of the misclassification error rate and
any real method will have a larger rate than this. This is a useful
approximation when you have a lot of individuals.</p>
    
    </div>

    <pre class="usage"><span class='fu'>oracle_mis</span>(<span class='no'>n</span>, <span class='no'>ploidy</span>, <span class='no'>seq</span>, <span class='no'>bias</span>, <span class='no'>od</span>, <span class='no'>dist</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>n</th>
      <td><p>The read-depth.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>The ploidy of the individual.</p></td>
    </tr>
    <tr>
      <th>seq</th>
      <td><p>The sequencing error rate.</p></td>
    </tr>
    <tr>
      <th>bias</th>
      <td><p>The allele-bias.</p></td>
    </tr>
    <tr>
      <th>od</th>
      <td><p>The overdispersion parameter.</p></td>
    </tr>
    <tr>
      <th>dist</th>
      <td><p>The distribution of the alleles.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A double. The oracle misclassification error rate.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>To come up with <code>dist</code>, you need some additional assumptions.
For example, if the population is in Hardy-Weinberg equilibrium and
the allele frequency is <code>alpha</code> then you could calculate
<code>dist</code> using the R code: <code>dbinom(x = 0:ploidy, size = ploidy, prob = alpha)</code>.
Alternatively, if you know the genotypes of the individual's two parents are, say,
<code>ref_count1</code> and <code>ref_count2</code>, then you could use the <code><a href='get_q_array.html'>get_q_array</a></code>
function from the updog package: <code>get_q_array(ploidy)[ref_count1 + 1, ref_count2 + 1, ]</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Gerard, David, Luis Felipe Ventorim Ferrao,
Antonio Augusto Franco Garcia, and Matthew Stephens. 2018.
Harnessing Empirical Bayes and Mendelian Segregation
for Genotyping Autopolyploids from Messy Sequencing Data."
<em>bioRxiv</em>. Cold Spring Harbor Laboratory. doi:10.1101/281550.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Hardy-Weinberg population with allele-frequency of 0.75.</span>
<span class='co'>## Moderate bias and moderate overdispersion.</span>
<span class='co'>## See how oracle misclassification error rates change as we</span>
<span class='co'>## increase the ploidy.</span>
<span class='no'>ploidy</span> <span class='kw'>&lt;-</span> <span class='fl'>2</span>
<span class='no'>dist</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/stats/topics/Binomial'>dbinom</a></span>(<span class='fl'>0</span>:<span class='no'>ploidy</span>, <span class='no'>ploidy</span>, <span class='fl'>0.75</span>)
<span class='fu'>oracle_mis</span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>ploidy</span>, <span class='kw'>seq</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
           <span class='kw'>bias</span> <span class='kw'>=</span> <span class='fl'>0.7</span>, <span class='kw'>od</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>dist</span> <span class='kw'>=</span> <span class='no'>dist</span>)</div><div class='output co'>#&gt; [1] 1.262647e-06</div><div class='input'>
<span class='no'>ploidy</span> <span class='kw'>&lt;-</span> <span class='fl'>4</span>
<span class='no'>dist</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/stats/topics/Binomial'>dbinom</a></span>(<span class='fl'>0</span>:<span class='no'>ploidy</span>, <span class='no'>ploidy</span>, <span class='fl'>0.75</span>)
<span class='fu'>oracle_mis</span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>ploidy</span>, <span class='kw'>seq</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
           <span class='kw'>bias</span> <span class='kw'>=</span> <span class='fl'>0.7</span>, <span class='kw'>od</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>dist</span> <span class='kw'>=</span> <span class='no'>dist</span>)</div><div class='output co'>#&gt; [1] 0.02944818</div><div class='input'>
<span class='no'>ploidy</span> <span class='kw'>&lt;-</span> <span class='fl'>6</span>
<span class='no'>dist</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/stats/topics/Binomial'>dbinom</a></span>(<span class='fl'>0</span>:<span class='no'>ploidy</span>, <span class='no'>ploidy</span>, <span class='fl'>0.75</span>)
<span class='fu'>oracle_mis</span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>ploidy</span>, <span class='kw'>seq</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
           <span class='kw'>bias</span> <span class='kw'>=</span> <span class='fl'>0.7</span>, <span class='kw'>od</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>dist</span> <span class='kw'>=</span> <span class='no'>dist</span>)</div><div class='output co'>#&gt; [1] 0.1329197</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

