<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Simulate individual genotypes from one of the supported <code><a href='flexdog.html'>flexdog</a></code> models. — rgeno • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Simulate individual genotypes from one of the supported <code><a href='flexdog.html'>flexdog</a></code> models. — rgeno" />

<meta property="og:description" content="This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
flexdog. See the details of flexdog for the models allowed." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate individual genotypes from one of the supported <code><a href='flexdog.html'>flexdog</a></code> models.</h1>
    
    <div class="hidden name"><code>rgeno.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
<code><a href='flexdog.html'>flexdog</a></code>. See the details of <code><a href='flexdog.html'>flexdog</a></code> for the models allowed.</p>
    
    </div>

    <pre class="usage"><span class='fu'>rgeno</span>(<span class='no'>n</span>, <span class='no'>ploidy</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"hw"</span>, <span class='st'>"bb"</span>, <span class='st'>"norm"</span>, <span class='st'>"ash"</span>, <span class='st'>"f1"</span>, <span class='st'>"s1"</span>, <span class='st'>"f1pp"</span>,
  <span class='st'>"s1pp"</span>, <span class='st'>"flex"</span>, <span class='st'>"uniform"</span>), <span class='kw'>allele_freq</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>od</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>p1geno</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p2geno</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>mode</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>pivec</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>mu</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sigma</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p1_pair_weights</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>p2_pair_weights</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>n</th>
      <td><p>The number of observations.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>The ploidy of the species.</p></td>
    </tr>
    <tr>
      <th>model</th>
      <td><p>What form should the prior take? See Details in <code><a href='flexdog.html'>flexdog</a></code>.</p></td>
    </tr>
    <tr>
      <th>allele_freq</th>
      <td><p>If <code>model = "hw"</code>, then this is the allele frequency of the population.
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>od</th>
      <td><p>If <code>model = "bb"</code>, then this is the overdispersion parameter of the beta-binomial
distribution. See <code><a href='betabinom.html'>betabinom</a></code> for details. For any other model, this should be
<code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>p1geno</th>
      <td><p>Either the first parent's genotype if <code>model = "f1"</code>
(or <code>model = "f1pp"</code>), or the only parent's
genotype if <code>model = "s1"</code> (or <code>model = "s1pp"</code>).
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>p2geno</th>
      <td><p>The second parent's genotype if <code>model = "f1"</code>
(or <code>model = "f1pp"</code>).
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>mode</th>
      <td><p>If <code>model = "ash"</code>, this is the center of the distribution. This should be a non-integer value
(so the mode is either the floor or the ceiling of <code>mode</code>).
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>pivec</th>
      <td><p>A vector of probabilities. If <code>model = "ash"</code>, then this represents
the mixing proportions of the discrete uniforms. If
<code>model = "flex"</code>, then element <code>i</code> is the probability of genotype <code>i - 1</code>.
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>mu</th>
      <td><p>If <code>model = "norm"</code>, this is the mean of the normal.
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>sigma</th>
      <td><p>If <code>model = "norm"</code>, this is the standard deviation of the normal.
For any other model, this should be <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>p1_pair_weights</th>
      <td><p>The mixing weights for the bivalent
pairs output in <code><a href='get_bivalent_probs.html'>get_bivalent_probs</a></code>
at the <code>lvec</code> level of <code>p1geno</code>.
If <code>model = "f1pp"</code> then this is for the first parent.
If <code>model = "s1pp"</code>, then this is for the only parent.
This should be <code>NULL</code> for all other values of <code>model</code>.</p></td>
    </tr>
    <tr>
      <th>p2_pair_weights</th>
      <td><p>If <code>model = "s1pp"</code>,
these are the mixing weights for the bivalent
pairs output in <code><a href='get_bivalent_probs.html'>get_bivalent_probs</a></code> at the
<code>lvec</code> level of <code>p2geno</code> for the second
parent.
This should be <code>NULL</code> for all other values of <code>model</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A vector of length <code>n</code> with the genotypes of the sampled individuals.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The allowable variable values of <code>allele_freq</code>, <code>od</code>, <code>p1geno</code>, <code>p2geno</code>, <code>pivec</code>,
and <code>mode</code> varies based on the value of <code>model</code>. If <code>model = "ash"</code>, then only
<code>mode</code> and <code>pivec</code> can be non-<code>NULL</code>. If <code>model = "flex"</code> then only
<code>pivec</code> can be non-<code>NULL</code>. If <code>model = "hw"</code>, then only <code>allele_freq</code> can
be non-<code>NULL</code>. If <code>model = "f1"</code> then only <code>p1geno</code> and <code>p2geno</code> can be non-<code>NULL</code>.
If <code>model = "s1"</code> then only <code>p1geno</code> can be non-<code>NULL</code>. If <code>model = "uniform"</code>, then
none of the above variables can be non-<code>NULL</code>. If <code>model = "bb"</code>, then only <code>allele_freq</code>,
and <code>od</code> can be non-<code>NULL</code>. If <code>model == "norm"</code>, then only <code>mu</code> and <code>sigma</code>
can be non-<code>NULL</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Gerard, D., Ferrao, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. *Genetics*, 210(3), 789-807. doi: [10.1534/genetics.118.301468](https://doi.org/10.1534/genetics.118.301468).</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## F1 Population where parent 1 has 1 copy of the referenc allele</span>
<span class='co'>## and parent 2 has 4 copies of the reference allele.</span>
<span class='no'>ploidy</span> <span class='kw'>&lt;-</span> <span class='fl'>6</span>
<span class='fu'>rgeno</span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>ploidy</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='st'>"f1"</span>, <span class='kw'>p1geno</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>p2geno</span> <span class='kw'>=</span> <span class='fl'>4</span>)</div><div class='output co'>#&gt;  [1] 3 3 3 2 2 4 1 2 2 2</div><div class='input'>
<span class='co'>## A population in Hardy-Weinberge equilibrium with an</span>
<span class='co'>## allele frequency of 0.75</span>
<span class='fu'>rgeno</span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='no'>ploidy</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='st'>"hw"</span>, <span class='kw'>allele_freq</span> <span class='kw'>=</span> <span class='fl'>0.75</span>)</div><div class='output co'>#&gt;  [1] 5 3 3 4 3 5 5 4 4 4</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

