<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title><code>updog</code> Flexible Genotyping for Polyploids — updog • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="<code>updog</code> Flexible Genotyping for Polyploids — updog" />

<meta property="og:description" content="Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allelic bias, overdispersion, and sequencing
error. The main function is flexdog, which allows
the specification of many different genotype distributions.
An experimental function that takes into account varying
levels of relatedness is implemented in mupdog.
Also provided are functions to simulate genotypes
(rgeno) and read-counts
(rflexdog), as well as functions to calculate
oracle genotyping error rates (oracle_mis) and
correlation with the true genotypes (oracle_cor).
These latter two functions are useful for read depth calculations.
Run browseVignettes(package = &quot;updog&quot;) in R
for example usage." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/parallel_computing.html">Parallelized Genotyping using Updog</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>updog</code> Flexible Genotyping for Polyploids</h1>
    
    <div class="hidden name"><code>updog.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allelic bias, overdispersion, and sequencing
error. The main function is <code><a href='flexdog.html'>flexdog</a></code>, which allows
the specification of many different genotype distributions.
An experimental function that takes into account varying
levels of relatedness is implemented in <code><a href='mupdog.html'>mupdog</a></code>.
Also provided are functions to simulate genotypes
(<code><a href='rgeno.html'>rgeno</a></code>) and read-counts
(<code><a href='rflexdog.html'>rflexdog</a></code>), as well as functions to calculate
oracle genotyping error rates (<code><a href='oracle_mis.html'>oracle_mis</a></code>) and
correlation with the true genotypes (<code><a href='oracle_cor.html'>oracle_cor</a></code>).
These latter two functions are useful for read depth calculations.
Run <code>browseVignettes(package = "updog")</code> in R
for example usage.</p>
    
    </div>

        
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The package is named <code>updog</code> for "Using
Parental Data for Offspring Genotyping" because
we originally developed the
method for full-sib populations, but it works
now for more general populations.</p>
<p>Our best competitor is probably the <code>fitPoly</code> package,
which you can check out at
<a href='https://cran.r-project.org/package=fitPoly'>https://cran.r-project.org/package=fitPoly</a>. Though, we think
that <code>updog</code> returns better calibrated measures of uncertainty
when you have next-generation sequencing data.</p>
<p>If you find a bug or want an enhancement, please submit an
issue at <a href='http://github.com/dcgerard/updog/issues'>http://github.com/dcgerard/updog/issues</a>.</p>
    
    <h2 class="hasAnchor" id="updog-functions"><a class="anchor" href="#updog-functions"></a><code>updog</code> Functions</h2>

    
    <dl class='dl-horizontal'>
  <dt><code><a href='flexdog.html'>flexdog</a></code></dt><dd><p>The main function that
      fits an empirical Bayes approach to genotype polyploids
      from next generation sequencing data.</p></dd>
  <dt><code><a href='mupdog.html'>mupdog</a></code></dt><dd><p>An experimental approach to genotype
      autopolyploids that accounts for varying levels of
      relatedness between the individuals in the sample.</p></dd>
  <dt><code><a href='rgeno.html'>rgeno</a></code></dt><dd><p>simulate the genotypes of a sample
      from one of the models allowed in <code><a href='flexdog.html'>flexdog</a></code>.</p></dd>
  <dt><code><a href='rflexdog.html'>rflexdog</a></code></dt><dd><p>Simulate read-counts from the
      <code><a href='flexdog.html'>flexdog</a></code> model.</p></dd>
  <dt><code><a href='plot.flexdog.html'>plot.flexdog</a></code></dt><dd><p>Plotting the output of
      <code><a href='flexdog.html'>flexdog</a></code>.</p></dd>
  <dt><code><a href='plot.mupdog.html'>plot.mupdog</a></code></dt><dd><p>Plotting the output of
      <code><a href='mupdog.html'>mupdog</a></code>.</p></dd>
  <dt><code><a href='oracle_joint.html'>oracle_joint</a></code></dt><dd><p>The joint distribution of the true genotype
      and an oracle estimator.</p></dd>
  <dt><code><a href='oracle_plot.html'>oracle_plot</a></code></dt><dd><p>Visualize the output of <code><a href='oracle_joint.html'>oracle_joint</a></code>.</p></dd>
  <dt><code><a href='oracle_mis.html'>oracle_mis</a></code></dt><dd><p>The oracle misclassification error rate (Bayes rate).</p></dd>
  <dt><code><a href='oracle_cor.html'>oracle_cor</a></code></dt><dd><p>Correlation between the true genotype and the oracle estimated genotype.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="updog-datasets"><a class="anchor" href="#updog-datasets"></a><code>updog</code> Datasets</h2>

    
    <dl class='dl-horizontal'>
  <dt><code><a href='snpdat.html'>snpdat</a></code></dt><dd><p>A small example dataset for using
      <code><a href='flexdog.html'>flexdog</a></code>.</p></dd>
  <dt><code><a href='uitdewilligen.html'>uitdewilligen</a></code></dt><dd><p>A small example dataset
      for using <code><a href='mupdog.html'>mupdog</a></code>.</p></dd>
  <dt><code><a href='mupout.html'>mupout</a></code></dt><dd><p>The output from fitting
      <code><a href='mupdog.html'>mupdog</a></code> to <code><a href='uitdewilligen.html'>uitdewilligen</a></code>.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Gerard, D., Ferrao, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. *Genetics*, 210(3), 789-807. doi: [10.1534/genetics.118.301468](https://doi.org/10.1534/genetics.118.301468).</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#details">Details</a></li>

      <li><a href="#updog-functions"><code>updog</code> Functions</a></li>

      <li><a href="#updog-datasets"><code>updog</code> Datasets</a></li>

      <li><a href="#references">References</a></li>
          </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

