# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Density function of betabinomial with the shape parameterizations
#'
#' @inheritParams dbetabinom_double
#' @param alpha The first shape parameter.
#' @param beta The second shape paramter.
#'
#' @author David Gerard
dbetabinom_alpha_beta_double <- function(x, size, alpha, beta, log) {
    .Call('_mupdog_dbetabinom_alpha_beta_double', PACKAGE = 'mupdog', x, size, alpha, beta, log)
}

#' Special case of betabinomial where the beta is bernoulli mu.
#'
#' @inheritParams dbetabinom_double
#'
#' @author David Gerard
dbernbinom <- function(x, size, mu, log) {
    .Call('_mupdog_dbernbinom', PACKAGE = 'mupdog', x, size, mu, log)
}

#' The density function of the beta-binomial distribution.
#'
#' @param x The quantile.
#' @param size The total number of draws.
#' @param mu The mean of the beta.
#' @param rho The overdispersion parameter of the beta.
#' @param log A logical. Should we return the log of the
#'     density \code{TRUE} or not \code{FALSE}?
#'
#' @author David Gerard
dbetabinom_double <- function(x, size, mu, rho, log) {
    .Call('_mupdog_dbetabinom_double', PACKAGE = 'mupdog', x, size, mu, rho, log)
}

#' @describeIn betabinom Density function.
#'
#' @export
#'
dbetabinom <- function(x, size, mu, rho, log) {
    .Call('_mupdog_dbetabinom', PACKAGE = 'mupdog', x, size, mu, rho, log)
}

#' The distribution function of the betabinomial. This is generally
#' only adviseable if q is relatively small.
#'
#' @inheritParams dbetabinom_double
#' @param q A quantile.
#' @param log_p A logical. Should return the log-probability
#'     \code{TRUE} or not \code{FALSE}?
#'
#' @author David Gerard
pbetabinom_double <- function(q, size, mu, rho, log_p) {
    .Call('_mupdog_pbetabinom_double', PACKAGE = 'mupdog', q, size, mu, rho, log_p)
}

#' @describeIn betabinom Distribution function.
#'
#' @export
#'
pbetabinom <- function(q, size, mu, rho, log_p) {
    .Call('_mupdog_pbetabinom', PACKAGE = 'mupdog', q, size, mu, rho, log_p)
}

#' The quantile function of the beta-binomial distribution parameterized
#' by mean and overdispersion parameter.
#'
#' @inheritParams dbetabinom_double
#' @param p The lower tail probability.
#'
#' @author David Gerard
qbetabinom_double <- function(p, size, mu, rho) {
    .Call('_mupdog_qbetabinom_double', PACKAGE = 'mupdog', p, size, mu, rho)
}

#' @describeIn betabinom Distribution function.
#'
#' @export
#'
qbetabinom <- function(p, size, mu, rho) {
    .Call('_mupdog_qbetabinom', PACKAGE = 'mupdog', p, size, mu, rho)
}

#' One draw from the beta-binomial distribution parameterized
#' by mean and overdispersion parameter.
#'
#' @inheritParams dbetabinom_double
#'
#' @author David Gerard
rbetabinom_int <- function(size, mu, rho) {
    .Call('_mupdog_rbetabinom_int', PACKAGE = 'mupdog', size, mu, rho)
}

#' @describeIn betabinom Distribution function.
#'
#' @export
#'
rbetabinom <- function(n, size, mu, rho) {
    .Call('_mupdog_rbetabinom', PACKAGE = 'mupdog', n, size, mu, rho)
}

#' Gradient for \code{\link{obj_for_mu_sigma2}} with respect for \code{mu} and \code{sigma2}.
#'
#' @inheritParams obj_for_mu_sigma2
#'
#' @author David Gerard
grad_for_mu_sigma2 <- function(mu, sigma2, phifk_mat, cor_inv, log_bb_dense) {
    .Call('_mupdog_grad_for_mu_sigma2', PACKAGE = 'mupdog', mu, sigma2, phifk_mat, cor_inv, log_bb_dense)
}

#' Gradient for \code{\link{obj_for_mu_sigma2_wrapper}} with respect for \code{muSigma2}
#' and a wrapper for \code{\link{grad_for_mu_sigma2}}
#'
#' @inheritParams obj_for_mu_sigma2
#' @param muSigma2 A vector. The first half are mu and the second half are sigma2.
#'
#' @author David Gerard
grad_for_mu_sigma2_wrapper <- function(muSigma2, phifk_mat, cor_inv, log_bb_dense) {
    .Call('_mupdog_grad_for_mu_sigma2_wrapper', PACKAGE = 'mupdog', muSigma2, phifk_mat, cor_inv, log_bb_dense)
}

#' Derivative of \deqn{-log(h) - (log(h) - \mu_h)^2 / (2\sigma_h^2)} with respect
#' to \eqn{h}.
#'
#' @param h The current bias parameter.
#' @param mu_h The mean of the log-bias.
#' @param sigma2_h The variance of the log-bias.
#'
#'
#' @seealso \code{\link{pen_bias}} which this is a derivative for.
#'
#' @author David Gerard
dpen_dh <- function(h, mu_h, sigma2_h) {
    .Call('_mupdog_dpen_dh', PACKAGE = 'mupdog', h, mu_h, sigma2_h)
}

#' Derivative of \deqn{-log(\epsilon(1 - \epsilon)) - (logit(\epsilon) - \mu_{\epsilon})^2 / (2\sigma_{\epsilon}^2)}
#' with respect to \eqn{\epsilon}.
#'
#' @param eps The current sequencing error rate.
#' @param mu_eps The mean of the logit of the sequencing error rate.
#' @param sigma2_eps The variance of the logit of the sequencing error rate.
#'
#' @seealso \code{\link{pen_seq_error}} which this is a derivative for.
#'
#' @author David Gerard
dpen_deps <- function(eps, mu_eps, sigma2_eps) {
    .Call('_mupdog_dpen_deps', PACKAGE = 'mupdog', eps, mu_eps, sigma2_eps)
}

#' Derivative of the log-beta density with
#' respect to c where \eqn{c = (1 - \tau)/\tau}
#' where \eqn{\tau} is the overdispersion parameter.
#'
#' @param x The number of successes observed
#' @param n The total number of trials observed.
#' @param xi The mean of the beta-binomial.
#' @param c \eqn{(1 - \tau)/\tau} where \eqn{\tau}
#'     is the overdispersion parameter.
#'
#'
#' @seealso \code{\link{dbetabinom_double}}, \code{\link{dlbeta_dtau}},
#'     \code{\link{dc_dtau}}.
#' @author David Gerard
dlbeta_dc <- function(x, n, xi, c) {
    .Call('_mupdog_dlbeta_dc', PACKAGE = 'mupdog', x, n, xi, c)
}

#' Derivative of \eqn{c = (1 - \tau) / \tau} with respect to \eqn{\tau}.
#'
#' @param tau The overdispersion parameter.
#'
#' @seealso \code{\link{dlbeta_dc}}, \code{\link{dlbeta_dtau}}
#'
#' @author David Gerard
dc_dtau <- function(tau) {
    .Call('_mupdog_dc_dtau', PACKAGE = 'mupdog', tau)
}

#' Derivative of the log-beta-binomial density with respect to the
#' overdispersion parameter.
#'
#' @param x The number of successes.
#' @param n The number of trials.
#' @param tau The overdispersion parameter.
#' @param p The allele dosage.
#' @param eps The sequencing error rate
#' @param h The bias parameter.
#'
#' @seealso \code{\link{dlbeta_dc}}, \code{\link{dc_dtau}},
#'     \code{\link{dbetabinom_double}}.
#'
#' @author David Gerard
dlbeta_dtau <- function(x, n, p, eps, h, tau) {
    .Call('_mupdog_dlbeta_dtau', PACKAGE = 'mupdog', x, n, p, eps, h, tau)
}

#' Derivative of the log-betabinomial density with respect to the
#' mean of the underlying beta.
#'
#' @inheritParams dlbeta_dtau
#' @param xi The mean of the underlying beta.
#'
#' @author David Gerard
dlbeta_dxi <- function(x, n, xi, tau) {
    .Call('_mupdog_dlbeta_dxi', PACKAGE = 'mupdog', x, n, xi, tau)
}

#' Derivative of xi-function with respect to bias parameter.
#'
#' @param p The dosage (between 0 and 1).
#' @param eps The sequencing error rate.
#' @param h The bias parameter.
#'
#' @author David Gerard
dxi_dh <- function(p, eps, h) {
    .Call('_mupdog_dxi_dh', PACKAGE = 'mupdog', p, eps, h)
}

#' Derivative of log-betabinomial density with respect to bias parameter.
#'
#' @inheritParams dlbeta_dtau
#'
#' @author David Gerard
dlbeta_dh <- function(x, n, p, eps, h, tau) {
    .Call('_mupdog_dlbeta_dh', PACKAGE = 'mupdog', x, n, p, eps, h, tau)
}

#' Derivative of xi with respect to f.
#'
#' @param h The bias parameter.
#' @param f The post-sequencing error rate adjusted probability of an A.
#'
#' @author David Gerard
dxi_df <- function(h, f) {
    .Call('_mupdog_dxi_df', PACKAGE = 'mupdog', h, f)
}

#' Derivative of f with respect to eps.
#'
#' @param p The allele dosage.
#' @param eps The sequencing error rate.
#'
#' @author David Gerard
df_deps <- function(p, eps) {
    .Call('_mupdog_df_deps', PACKAGE = 'mupdog', p, eps)
}

#' Derivative of the log-beta-binomial density with respect to the
#' sequencing error rate.
#'
#' @inheritParams dlbeta_dtau
#'
#' @author David Gerard
dlbeta_deps <- function(x, n, p, eps, h, tau) {
    .Call('_mupdog_dlbeta_deps', PACKAGE = 'mupdog', x, n, p, eps, h, tau)
}

#' Gradient for \code{\link{obj_for_eps}}.
#'
#' @inheritParams obj_for_eps
#'
#' @author David Gerard
grad_for_eps <- function(parvec, refvec, sizevec, ploidy, mean_bias, var_bias, mean_seq, var_seq, wmat) {
    .Call('_mupdog_grad_for_eps', PACKAGE = 'mupdog', parvec, refvec, sizevec, ploidy, mean_bias, var_bias, mean_seq, var_seq, wmat)
}

#' Variational posterior probability of having \code{dosage} A alleles
#' when the ploidy is \code{ploidy}, the allele frequency is
#' \code{alpha}, the individual-specific overdispersion parameter is
#' \code{rho}, the variational mean is \code{mu}, and the variational
#' variance is \code{sigma2}.
#'
#' @param dosage The number of A alleles.
#' @param ploidy The ploidy of the individual.
#' @param mu The variational mean.
#' @param sigma2 The variational variance (not standard devation).
#' @param alpha The allele frequency.
#' @param rho The individual's overdispersion parameter.
#'
#' @author David
#'
post_prob <- function(dosage, ploidy, mu, sigma2, alpha, rho) {
    .Call('_mupdog_post_prob', PACKAGE = 'mupdog', dosage, ploidy, mu, sigma2, alpha, rho)
}

#' Computes every posterior probability for each dosage level for
#' each individual at each SNP.
#'
#' @param ploidy The ploidy of the species.
#' @param mu A matrix of variational posterior means. The rows
#'     index the individuals and the columns index the SNPs.
#' @param sigma2 A matrix of variational posterior variances.
#'     The rows index the individuals and the columns index the SNPs.
#' @param alpha A vector of allele frequencies for all SNPs.
#' @param rho A vector of inbreeding coefficients for all individuals.
#'
#' @author David Gerard
#'
compute_all_post_prob <- function(ploidy, mu, sigma2, alpha, rho) {
    .Call('_mupdog_compute_all_post_prob', PACKAGE = 'mupdog', ploidy, mu, sigma2, alpha, rho)
}

#' Calculates the log-density for every individual by snp by dosage level.
#'
#' @inheritParams mupdog
#'
#'
#'
compute_all_log_bb <- function(refmat, sizemat, ploidy, seq, bias, od) {
    .Call('_mupdog_compute_all_log_bb', PACKAGE = 'mupdog', refmat, sizemat, ploidy, seq, bias, od)
}

#' Computes \deqn{\Phi^{-1}(F(k|K,\alpha_j,\rho_i))} for all possible (i,j,k).
#'
#' @param alpha A vector whose jth element is the allele frequency of SNP j.
#' @param rho A vector whose ith element is the inbreeding coefficient of individual i.
#' @param ploidy The ploidy of the species.
#'
#' @author David Gerard
compute_all_phifk <- function(alpha, rho, ploidy) {
    .Call('_mupdog_compute_all_phifk', PACKAGE = 'mupdog', alpha, rho, ploidy)
}

#' Penalty on bias parameter.
#'
#' @param h The current value of bias parameter. Must be
#'     greater than 0. A value of 1 means no bias.
#' @param mu_h The prior mean of the log-bias parameter.
#' @param sigma2_h The prior variance (not standard deviation)
#'     of the log-bias parameter.
#'
#' @author David Gerard
pen_bias <- function(h, mu_h, sigma2_h) {
    .Call('_mupdog_pen_bias', PACKAGE = 'mupdog', h, mu_h, sigma2_h)
}

#' Penalty on sequencing error rate.
#'
#' @param eps The current value of sequencing error rate.
#'     Must be between 0 and 1.
#' @param mu_eps The prior mean of the logit sequencing error rate.
#' @param sigma2_eps The prior variance (not standard deviation)
#'     of the logit sequencing error rate.
#'
#' @author David Gerard
pen_seq_error <- function(eps, mu_eps, sigma2_eps) {
    .Call('_mupdog_pen_seq_error', PACKAGE = 'mupdog', eps, mu_eps, sigma2_eps)
}

#' Objective function when updating a single inbreeding coefficient.
#'
#' @param rho The inbreeding coefficient for the individual.
#' @param mu A vector of posterior means. The jth element is the
#'     posterior mean of SNP j for the individual.
#' @param sigma2 A vector of posterior variances. The jth element
#'     is the posterior variance of SNP j for the individual.
#' @param alpha A vector of allele frequencies. The jth element
#'     is the allele frequency for SNP j.
#' @param log_bb_dense A matrix of log posterior densities. The
#'     rows index the SNPs and the columns index the dosage.
#' @param ploidy The ploidy of the species.
#'
#'
#' @author David Gerard
obj_for_rho <- function(rho, mu, sigma2, alpha, log_bb_dense, ploidy) {
    .Call('_mupdog_obj_for_rho', PACKAGE = 'mupdog', rho, mu, sigma2, alpha, log_bb_dense, ploidy)
}

#' Objective function when updating alpha
#'
#' @param mu A vector. The ith element is individual i's variational posterior mean at the SNP.
#' @param sigma2 A vector. The ith element is individual i's variational posterior variance at the SNP.
#' @param alpha The SNP's allele frequency.
#' @param rho A vector. The ith element is individuals i's inbreeding coefficient.
#' @param log_bb_dense A matrix of log-densities of the beta binomial. The rows index the individuals and the columns index the allele dosage.
#' @param ploidy The ploidy of the species.
#'
#'
#'
#' @author David Gerard
obj_for_alpha <- function(mu, sigma2, alpha, rho, log_bb_dense, ploidy) {
    .Call('_mupdog_obj_for_alpha', PACKAGE = 'mupdog', mu, sigma2, alpha, rho, log_bb_dense, ploidy)
}

#' Objective function for updating sequencing error rate, bias, and overdispersion parameters.
#'
#' @param parvec A vector of length three. The first element is the sequencing
#'     error rate, the second element is the allele bias, and the third element
#'     is the overdispersion parameter.
#' @param refvec A vector. The ith element is the reference count for the ith individual in the SNP.
#' @param sizevec A vector. the ith element is the size count for the ith individual in the SNP/
#' @param ploidy The ploidy of the species.
#' @param mean_bias The prior mean of the log-bias.
#' @param var_bias The prior variance of the log-bias
#' @param mean_seq The prior mean of the logit sequencing error rate.
#' @param var_seq The prior variance of the logit sequencing error rate.
#' @param wmat The matrix of variational posterior probabilities for each dosage.
#'     The rows index the individuals and the columns index the dosage levels.
#'
#' @author David Gerard
obj_for_eps <- function(parvec, refvec, sizevec, ploidy, mean_bias, var_bias, mean_seq, var_seq, wmat) {
    .Call('_mupdog_obj_for_eps', PACKAGE = 'mupdog', parvec, refvec, sizevec, ploidy, mean_bias, var_bias, mean_seq, var_seq, wmat)
}

#' Objective function when updating mu and sigma2.
#'
#' @param mu A vector, the ith element is the variational posterior mean of individual i for the SNP.
#' @param sigma2 A vector, the ith element is the variational posterior variance of individual i for the SNP.
#' @param phifk_mat A matrix that contains the standard normal quantile of the beta-binomial cdf at dosage k for individual i.
#'     The rows index the individuals and the columns index the dosages.
#' @param cor_inv The inverse of the underlying correlation matrix.
#' @param log_bb_dense A matrix of log-densities of the beta binomial. The rows index the individuals and the columns index the allele dosage.
#'     Allele dosage goes from -1 to ploidy, so there are ploidy + 2 elements.
#'
#'
#' @author David Gerard
obj_for_mu_sigma2 <- function(mu, sigma2, phifk_mat, cor_inv, log_bb_dense) {
    .Call('_mupdog_obj_for_mu_sigma2', PACKAGE = 'mupdog', mu, sigma2, phifk_mat, cor_inv, log_bb_dense)
}

#' Wrapper for \code{\link{obj_for_mu_sigma2}} so that I can use it in \code{optim}.
#'
#' @inheritParams obj_for_mu_sigma2
#' @param muSigma2 A vector. The first half are mu and the second half are sigma2.
#'
#'
#' @author David Gerard
obj_for_mu_sigma2_wrapper <- function(muSigma2, phifk_mat, cor_inv, log_bb_dense) {
    .Call('_mupdog_obj_for_mu_sigma2_wrapper', PACKAGE = 'mupdog', muSigma2, phifk_mat, cor_inv, log_bb_dense)
}

#' The evidence lower bound
#'
#' @param warray An three-way array. The (i,j,k)th entry is the variational posterior probability
#'     that individual i at SNP j has dosage k - 1. See \code{\link{compute_all_post_prob}}.
#' @param lbeta_array A three-way array. The (i,j,k)th entry is the log-density of the betabinomial
#'     for individual i at SNP j and dosage k - 1. See \code{\link{compute_all_log_bb}}.
#' @param cor_inv The inverse of the correlation matrix.
#' @param postmean A matrix. The (i,j)th entry is the variational posterior mean for individual i
#'     at SNP j.
#' @param postvar A matrix. The (i,j)th entry is the variational posterior variance for individual i at SNP j.
#' @param bias A vector. The jth entry is the allele bias for SNP j.
#' @param seq A vector. The jth entry is the sequencing error rate at SNP j.
#' @param mean_bias The prior mean on the log-bias.
#' @param var_bias The prior variance on the log-bias.
#' @param mean_seq The prior mean on the logit of the sequencing error rate.
#' @param var_seq The prior variance on the logit of the sequencing error rate.
#' @param ploidy The ploidy of the species.
#'
#'
#'
#'
#' @author David Gerard
elbo <- function(warray, lbeta_array, cor_inv, postmean, postvar, bias, seq, mean_bias, var_bias, mean_seq, var_seq, ploidy) {
    .Call('_mupdog_elbo', PACKAGE = 'mupdog', warray, lbeta_array, cor_inv, postmean, postvar, bias, seq, mean_bias, var_bias, mean_seq, var_seq, ploidy)
}

#' Calculate oracle misclassification error.
#'
#' Given knowledge of the parameters, we calculate the expected misclassification error,
#' where the expectation is taken over both the data generation process and the allele-distribution.
#' This is an ideal level of misclassification error.
#'
#' To come up with \code{dist}, you need some additional assumptions.
#' For example, if the population is in Hardy-Weinberg equilibrium and
#' the allele frequency is \code{alpha} then you could calculate
#' \code{dist} using the R code: \code{dbinom(x = 0:ploidy, size = ploidy, prob = alpha)}.
#' Alternatively, if you know the genotypes of the individual's two parents are, say,
#' \code{ref_count1} and \code{ref_count2}, then you could use the \code{\link[updog]{get_q_array}}
#' function from the updog package: \code{updog::get_q_array(ploidy)[ref_count1 + 1, ref_count2 + 1, ]}.
#'
#' @param n The read-depth.
#' @param ploidy The ploidy of the individual.
#' @param seq The sequencing error rate.
#' @param bias The allele-bias
#' @param od The overdispersion parameter.
#' @param dist The distribution of the alleles.
#'
#'
#' @author David Gerard
#'
#' @export
#'
oracle_miss <- function(n, ploidy, seq, bias, od, dist) {
    .Call('_mupdog_oracle_miss', PACKAGE = 'mupdog', n, ploidy, seq, bias, od, dist)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps}.
#'
#' @param p The allele dosage.
#' @param eps The sequencing error rate.
#'
#' @author David Gerard
eta_double <- function(p, eps) {
    .Call('_mupdog_eta_double', PACKAGE = 'mupdog', p, eps)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps}.
#'
#' @inheritParams xi_fun
#'
#' @author David Gerard
eta_fun <- function(p, eps) {
    .Call('_mupdog_eta_fun', PACKAGE = 'mupdog', p, eps)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps} and the allele bias \code{h}.
#'
#' @param p The allele dosage.
#' @param eps The sequencing error rate.
#' @param h The allele bias.
#'
#' @author David Gerard
xi_double <- function(p, eps, h) {
    .Call('_mupdog_xi_double', PACKAGE = 'mupdog', p, eps, h)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps} and the allele bias \code{h}.
#'
#' @param p A vector of allele dosages.
#' @param eps The sequencing error rate. Must either be of length 1
#'     or the same length as p.
#' @param h The allele bias. Must either be of length 1 or the same length
#'     as p.
#'
#' @author David Gerard
xi_fun <- function(p, eps, h) {
    .Call('_mupdog_xi_fun', PACKAGE = 'mupdog', p, eps, h)
}

#' Log-sum-exponential trick.
#'
#' @param x A vector to log-sum-exp.
#'
#' @return The log of the sum of the exponential
#'     of the elements in \code{x}.
#'
#' @author David Gerard
log_sum_exp <- function(x) {
    .Call('_mupdog_log_sum_exp', PACKAGE = 'mupdog', x)
}

#' Log-sum-exponential trick using just two doubles.
#'
#' @param x A double.
#' @param y Another double.
#'
#' @return The log of the sum of the exponential of x and y.
#'
#' @author David Gerard
#'
log_sum_exp_2 <- function(x, y) {
    .Call('_mupdog_log_sum_exp_2', PACKAGE = 'mupdog', x, y)
}

#' The logit function.
#'
#' @param x A double between 0 and 1.
#'
#' @author David Gerard
logit <- function(x) {
    .Call('_mupdog_logit', PACKAGE = 'mupdog', x)
}

#' The expit (logistic) function.
#'
#' @param x A double.
#'
#' @author David Gerard
expit <- function(x) {
    .Call('_mupdog_expit', PACKAGE = 'mupdog', x)
}

