# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Density function of betabinomial with the shape parameterizations
#'
#' @inheritParams dbetabinom_double
#' @param alpha The first shape parameter.
#' @param beta The second shape paramter.
#'
#' @author David Gerard
dbetabinom_alpha_beta_double <- function(x, size, alpha, beta, log) {
    .Call('_mupdog_dbetabinom_alpha_beta_double', PACKAGE = 'mupdog', x, size, alpha, beta, log)
}

#' Special case of betabinomial where the beta is bernoulli mu.
#'
#' @inheritParams dbetabinom_double
#'
#' @author David Gerard
dbernbinom <- function(x, size, mu, log) {
    .Call('_mupdog_dbernbinom', PACKAGE = 'mupdog', x, size, mu, log)
}

#' The density function of the beta-binomial distribution.
#'
#' @param x The quantile.
#' @param size The total number of draws.
#' @param mu The mean of the beta.
#' @param rho The overdispersion parameter of the beta.
#' @param log A logical. Should we return the log of the
#'     density \code{TRUE} or not \code{FALSE}?
#'
#' @author David Gerard
dbetabinom_double <- function(x, size, mu, rho, log) {
    .Call('_mupdog_dbetabinom_double', PACKAGE = 'mupdog', x, size, mu, rho, log)
}

#' @describeIn betabinom Density function.
dbetabinom <- function(x, size, mu, rho, log) {
    .Call('_mupdog_dbetabinom', PACKAGE = 'mupdog', x, size, mu, rho, log)
}

#' The distribution function of the betabinomial. This is generally
#' only adviseable if q is relatively small.
#'
#' @inheritParams dbetabinom_double
#' @param q A quantile.
#' @param log_p A logical. Should return the log-probability
#'     \code{TRUE} or not \code{FALSE}?
#'
#' @author David Gerard
pbetabinom_double <- function(q, size, mu, rho, log_p) {
    .Call('_mupdog_pbetabinom_double', PACKAGE = 'mupdog', q, size, mu, rho, log_p)
}

#' @describeIn betabinom Distribution function.
#'
pbetabinom <- function(q, size, mu, rho, log_p) {
    .Call('_mupdog_pbetabinom', PACKAGE = 'mupdog', q, size, mu, rho, log_p)
}

#' Variational posterior probability of having \code{dosage} A alleles
#' when the ploidy is \code{ploidy}, the allele frequency is
#' \code{alpha}, the individual-specific overdispersion parameter is
#' \code{rho}, the variational mean is \code{mu}, and the variational
#' variance is \code{sigma2}.
#'
#' @param dosage The number of A alleles.
#' @param ploidy The ploidy of the individual.
#' @param mu The variational mean.
#' @param sigma2 The variational variance (not standard devation).
#' @param alpha The allele frequency.
#' @param rho The individual's overdispersion parameter.
#'
#' @author David
#'
post_prob <- function(dosage, ploidy, mu, sigma2, alpha, rho) {
    .Call('_mupdog_post_prob', PACKAGE = 'mupdog', dosage, ploidy, mu, sigma2, alpha, rho)
}

#' Computes every posterior probability for each dosage level for
#' each individual at each SNP.
#'
#' @param ploidy The ploidy of the species.
#' @param mu A matrix of variational posterior means. The rows
#'     index the individuals and the columns index the SNPs.
#' @param sigma2 A matrix of variational posterior variances.
#'     The rows index the individuals and the columns index the SNPs.
#' @param alpha A vector of allele frequencies for all SNPs.
#' @param rho A vector of inbreeding coefficients for all individuals.
#'
#' @author David Gerard
#'
compute_all_post_prob <- function(ploidy, mu, sigma2, alpha, rho) {
    .Call('_mupdog_compute_all_post_prob', PACKAGE = 'mupdog', ploidy, mu, sigma2, alpha, rho)
}

#' Calculates the log-density for every individual by snp by dosage level.
#'
#' @inheritParams mupdog
#'
#'
#'
compute_all_log_bb <- function(refmat, sizemat, ploidy, seq, bias, od) {
    .Call('_mupdog_compute_all_log_bb', PACKAGE = 'mupdog', refmat, sizemat, ploidy, seq, bias, od)
}

#' Penalty on bias parameter.
#'
#' @param h The current value of bias parameter. Must be
#'     greater than 0. A value of 1 means no bias.
#' @param mu_h The prior mean of the log-bias parameter.
#' @param sigma2_h The prior variance (not standard deviation)
#'     of the log-bias parameter.
#'
#' @author David Gerard
pen_bias <- function(h, mu_h, sigma2_h) {
    .Call('_mupdog_pen_bias', PACKAGE = 'mupdog', h, mu_h, sigma2_h)
}

#' Penalty on sequencing error rate.
#'
#' @param eps The current value of sequencing error rate.
#'     Must be between 0 and 1.
#' @param mu_eps The prior mean of the logit sequencing error rate.
#' @param sigma2_eps The prior variance (not standard deviation)
#'     of the logit sequencing error rate.
#'
#' @author David Gerard
pen_seq_error <- function(eps, mu_eps, sigma2_eps) {
    .Call('_mupdog_pen_seq_error', PACKAGE = 'mupdog', eps, mu_eps, sigma2_eps)
}

#' Objective function when updating a single inbreeding coefficient.
#'
#' @param rho The inbreeding coefficient for the individual.
#' @param mu A vector of posterior means. The jth element is the
#'     posterior mean of SNP j for the individual.
#' @param sigma2 A vector of posterior variances. The jth element
#'     is the posterior variance of SNP j for the individual.
#' @param alpha A vector of allele frequencies. The jth element
#'     is the allele frequency for SNP j.
#' @param log_bb_dense A matrix of log posterior densities. The
#'     rows index the SNPs and the columns index the dosage.
#' @param ploidy The ploidy of the species.
#'
#'
#' @author David Gerard
obj_for_rho <- function(rho, mu, sigma2, alpha, log_bb_dense, ploidy) {
    .Call('_mupdog_obj_for_rho', PACKAGE = 'mupdog', rho, mu, sigma2, alpha, log_bb_dense, ploidy)
}

#' Objective function when updating mu, sigma2, and alpha
#'
#' @param mu A vector. The ith element is individual i's variational posterior mean at the SNP.
#' @param sigma2 A vector. The ith element is individual i's variational posterior variance at the SNP.
#' @param alpha The SNP's allele frequency.
#' @param rho A vector. The ith element is individuals i's inbreeding coefficient.
#' @param log_bb_dense A matrix of log-densities of the beta binomial. The rows index the individuals and the columns index the allele dosage.
#' @param ploidy The ploidy of the species.
#' @param cor_inv The inverse of the correlation matrix.
#'
#'
#'
#' @author David Gerard
obj_for_mu <- function(mu, sigma2, alpha, rho, log_bb_dense, ploidy, cor_inv) {
    .Call('_mupdog_obj_for_mu', PACKAGE = 'mupdog', mu, sigma2, alpha, rho, log_bb_dense, ploidy, cor_inv)
}

#' Wrapper for \code{\link{obj_for_mu}}.
#'
#' @inheritParams obj_for_mu
#' @param muSigma2Alpha A vector where the first \code{nsnps} observations are mu,
#'     the next \code{nsnps} observations are sigma2, and the last observation is \code{alpha}.
#'
#'
#'
#'
#'
#'
#' @author David Gerard
obj_for_mu_wrapper <- function(muSigma2Alpha, rho, log_bb_dense, ploidy, cor_inv) {
    .Call('_mupdog_obj_for_mu_wrapper', PACKAGE = 'mupdog', muSigma2Alpha, rho, log_bb_dense, ploidy, cor_inv)
}

#' Objective function for updating sequencing error rate, bias, and overdispersion parameters.
#'
#' @param parvec A vector of length three. The first element is the sequencing
#'     error rate, the second element is the allele bias, and the third element
#'     is the overdispersion parameter.
#' @param refvec A vector. The ith element is the reference count for the ith individual in the SNP.
#' @param sizevec A vector. the ith element is the size count for the ith individual in the SNP/
#' @param ploidy The ploidy of the species.
#' @param mean_bias The prior mean of the log-bias.
#' @param var_bias The prior variance of the log-bias
#' @param mean_seq The prior mean of the logit sequencing error rate.
#' @param var_seq The prior variance of the logit sequencing error rate.
#' @param wmat The matrix of variational posterior probabilities for each dosage.
#'     The rows index the individuals and the columns index the dosage levels.
#'
#' @author David Gerard
obj_for_eps <- function(parvec, refvec, sizevec, ploidy, mean_bias, var_bias, mean_seq, var_seq, wmat) {
    .Call('_mupdog_obj_for_eps', PACKAGE = 'mupdog', parvec, refvec, sizevec, ploidy, mean_bias, var_bias, mean_seq, var_seq, wmat)
}

#' The evidence lower bound
#'
#' @param warray An three-way array. The (i,j,k)th entry is the variational posterior probability
#'     that individual i at SNP j has dosage k - 1. See \code{\link{compute_all_post_prob}}.
#' @param lbeta_array A three-way array. The (i,j,k)th entry is the log-density of the betabinomial
#'     for individual i at SNP j and dosage k - 1. See \code{\link{compute_all_log_bb}}.
#' @param cor_inv The inverse of the correlation matrix.
#' @param postmean A matrix. The (i,j)th entry is the variational posterior mean for individual i
#'     at SNP j.
#' @param postvar A matrix. The (i,j)th entry is the variational posterior variance for individual i at SNP j.
#' @param bias A vector. The jth entry is the allele bias for SNP j.
#' @param seq A vector. The jth entry is the sequencing error rate at SNP j.
#' @param mean_bias The prior mean on the log-bias.
#' @param var_bias The prior variance on the log-bias.
#' @param mean_seq The prior mean on the logit of the sequencing error rate.
#' @param var_seq The prior variance on the logit of the sequencing error rate.
#' @param ploidy The ploidy of the species.
#'
#'
#'
#'
#' @author David Gerard
elbo <- function(warray, lbeta_array, cor_inv, postmean, postvar, bias, seq, mean_bias, var_bias, mean_seq, var_seq, ploidy) {
    .Call('_mupdog_elbo', PACKAGE = 'mupdog', warray, lbeta_array, cor_inv, postmean, postvar, bias, seq, mean_bias, var_bias, mean_seq, var_seq, ploidy)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps}.
#'
#' @param p The allele dosage.
#' @param eps The sequencing error rate.
#'
#' @author David Gerard
eta_double <- function(p, eps) {
    .Call('_mupdog_eta_double', PACKAGE = 'mupdog', p, eps)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps}.
#'
#' @inheritParams xi_fun
#'
#' @author David Gerard
eta_fun <- function(p, eps) {
    .Call('_mupdog_eta_fun', PACKAGE = 'mupdog', p, eps)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps} and the allele bias \code{h}.
#'
#' @param p The allele dosage.
#' @param eps The sequencing error rate.
#' @param h The allele bias.
#'
#' @author David Gerard
xi_double <- function(p, eps, h) {
    .Call('_mupdog_xi_double', PACKAGE = 'mupdog', p, eps, h)
}

#' Adjusts allele dosage \code{p} by the sequencing error rate \code{eps} and the allele bias \code{h}.
#'
#' @param p A vector of allele dosages.
#' @param eps The sequencing error rate. Must either be of length 1
#'     or the same length as p.
#' @param h The allele bias. Must either be of length 1 or the same length
#'     as p.
#'
#' @author David Gerard
xi_fun <- function(p, eps, h) {
    .Call('_mupdog_xi_fun', PACKAGE = 'mupdog', p, eps, h)
}

#' Log-sum-exponential trick.
#'
#' @param x A vector to log-sum-exp.
#'
#' @return The log of the sum of the exponential
#'     of the elements in \code{x}.
#'
#' @author David Gerard
log_sum_exp <- function(x) {
    .Call('_mupdog_log_sum_exp', PACKAGE = 'mupdog', x)
}

#' The logit function.
#'
#' @param x A double between 0 and 1.
#'
#' @author David Gerard
logit <- function(x) {
    .Call('_mupdog_logit', PACKAGE = 'mupdog', x)
}

#' The expit (logistic) function.
#'
#' @param x A double.
#'
#' @author David Gerard
expit <- function(x) {
    .Call('_mupdog_expit', PACKAGE = 'mupdog', x)
}

